{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2ffe5488",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:01:56.100881Z",
     "iopub.status.busy": "2023-10-10T12:01:56.100462Z",
     "iopub.status.idle": "2023-10-10T12:01:56.105676Z",
     "shell.execute_reply": "2023-10-10T12:01:56.104878Z"
    },
    "papermill": {
     "duration": 0.015907,
     "end_time": "2023-10-10T12:01:56.108741",
     "exception": false,
     "start_time": "2023-10-10T12:01:56.092834",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# scores_teamhydrogen-white-malamute-prompt-openorca-v2_test-cirrus-e5-large-title\n",
    "# 0.91941\n",
    "\n",
    "# scores_wealthy-ant_test-e5-base-title\n",
    "# 0.924605\n",
    "\n",
    "# scores_teamhydrogen-cool-giraffe-prompt-2409_test-gte-large-title\n",
    "# 0.92579\n",
    "\n",
    "# scores_teamhydrogen-nostalgic-mammoth-avg_test-cirrus-gte-base-title-3\n",
    "# 0.92602\n",
    "\n",
    "# scores_teamhydrogen-raspberry-beaver-1_test-cirrus-e5-large-title\n",
    "# 0.92659\n",
    "\n",
    "# scores_teamhydrogen-cool-giraffe-prompt-0710_test-cirrus-bge-large-title-ind\n",
    "# 0.92656"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4a25d64f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:01:56.121326Z",
     "iopub.status.busy": "2023-10-10T12:01:56.120797Z",
     "iopub.status.idle": "2023-10-10T12:02:02.942341Z",
     "shell.execute_reply": "2023-10-10T12:02:02.941326Z"
    },
    "papermill": {
     "duration": 6.830222,
     "end_time": "2023-10-10T12:02:02.944667",
     "exception": false,
     "start_time": "2023-10-10T12:01:56.114445",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from datasets import load_dataset\n",
    "from tqdm.notebook import tqdm\n",
    "import torch\n",
    "import pandas as pd\n",
    "from glob import glob\n",
    "import numpy as np\n",
    "import os\n",
    "os.environ[\"TRANSFORMERS_OFFLINE\"] = \"1\"\n",
    "os.environ[\"TOKENIZERS_PARALLELISM\"] = \"false\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "12897e81",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:02:02.959053Z",
     "iopub.status.busy": "2023-10-10T12:02:02.958519Z",
     "iopub.status.idle": "2023-10-10T12:02:40.299786Z",
     "shell.execute_reply": "2023-10-10T12:02:40.298597Z"
    },
    "papermill": {
     "duration": 37.351086,
     "end_time": "2023-10-10T12:02:40.302127",
     "exception": false,
     "start_time": "2023-10-10T12:02:02.951041",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing /kaggle/input/sci-llm-pip-v2/sentence-transformers-2.2.2.tar.gz\r\n",
      "  Preparing metadata (setup.py) ... \u001b[?25l-\b \b\\\b \bdone\r\n",
      "\u001b[?25hRequirement already satisfied: transformers<5.0.0,>=4.6.0 in /opt/conda/lib/python3.10/site-packages (from sentence-transformers==2.2.2) (4.30.2)\r\n",
      "Requirement already satisfied: tqdm in /opt/conda/lib/python3.10/site-packages (from sentence-transformers==2.2.2) (4.65.0)\r\n",
      "Requirement already satisfied: torch>=1.6.0 in /opt/conda/lib/python3.10/site-packages (from sentence-transformers==2.2.2) (2.0.0)\r\n",
      "Requirement already satisfied: torchvision in /opt/conda/lib/python3.10/site-packages (from sentence-transformers==2.2.2) (0.15.1)\r\n",
      "Requirement already satisfied: numpy in /opt/conda/lib/python3.10/site-packages (from sentence-transformers==2.2.2) (1.23.5)\r\n",
      "Requirement already satisfied: scikit-learn in /opt/conda/lib/python3.10/site-packages (from sentence-transformers==2.2.2) (1.2.2)\r\n",
      "Requirement already satisfied: scipy in /opt/conda/lib/python3.10/site-packages (from sentence-transformers==2.2.2) (1.11.1)\r\n",
      "Requirement already satisfied: nltk in /opt/conda/lib/python3.10/site-packages (from sentence-transformers==2.2.2) (3.2.4)\r\n",
      "Requirement already satisfied: sentencepiece in /opt/conda/lib/python3.10/site-packages (from sentence-transformers==2.2.2) (0.1.99)\r\n",
      "Requirement already satisfied: huggingface-hub>=0.4.0 in /opt/conda/lib/python3.10/site-packages (from sentence-transformers==2.2.2) (0.16.4)\r\n",
      "Requirement already satisfied: filelock in /opt/conda/lib/python3.10/site-packages (from huggingface-hub>=0.4.0->sentence-transformers==2.2.2) (3.12.2)\r\n",
      "Requirement already satisfied: fsspec in /opt/conda/lib/python3.10/site-packages (from huggingface-hub>=0.4.0->sentence-transformers==2.2.2) (2023.6.0)\r\n",
      "Requirement already satisfied: requests in /opt/conda/lib/python3.10/site-packages (from huggingface-hub>=0.4.0->sentence-transformers==2.2.2) (2.31.0)\r\n",
      "Requirement already satisfied: pyyaml>=5.1 in /opt/conda/lib/python3.10/site-packages (from huggingface-hub>=0.4.0->sentence-transformers==2.2.2) (6.0)\r\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in /opt/conda/lib/python3.10/site-packages (from huggingface-hub>=0.4.0->sentence-transformers==2.2.2) (4.6.3)\r\n",
      "Requirement already satisfied: packaging>=20.9 in /opt/conda/lib/python3.10/site-packages (from huggingface-hub>=0.4.0->sentence-transformers==2.2.2) (21.3)\r\n",
      "Requirement already satisfied: sympy in /opt/conda/lib/python3.10/site-packages (from torch>=1.6.0->sentence-transformers==2.2.2) (1.12)\r\n",
      "Requirement already satisfied: networkx in /opt/conda/lib/python3.10/site-packages (from torch>=1.6.0->sentence-transformers==2.2.2) (3.1)\r\n",
      "Requirement already satisfied: jinja2 in /opt/conda/lib/python3.10/site-packages (from torch>=1.6.0->sentence-transformers==2.2.2) (3.1.2)\r\n",
      "Requirement already satisfied: regex!=2019.12.17 in /opt/conda/lib/python3.10/site-packages (from transformers<5.0.0,>=4.6.0->sentence-transformers==2.2.2) (2023.6.3)\r\n",
      "Requirement already satisfied: tokenizers!=0.11.3,<0.14,>=0.11.1 in /opt/conda/lib/python3.10/site-packages (from transformers<5.0.0,>=4.6.0->sentence-transformers==2.2.2) (0.13.3)\r\n",
      "Requirement already satisfied: safetensors>=0.3.1 in /opt/conda/lib/python3.10/site-packages (from transformers<5.0.0,>=4.6.0->sentence-transformers==2.2.2) (0.3.1)\r\n",
      "Requirement already satisfied: six in /opt/conda/lib/python3.10/site-packages (from nltk->sentence-transformers==2.2.2) (1.16.0)\r\n",
      "Requirement already satisfied: joblib>=1.1.1 in /opt/conda/lib/python3.10/site-packages (from scikit-learn->sentence-transformers==2.2.2) (1.2.0)\r\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /opt/conda/lib/python3.10/site-packages (from scikit-learn->sentence-transformers==2.2.2) (3.1.0)\r\n",
      "Requirement already satisfied: pillow!=8.3.*,>=5.3.0 in /opt/conda/lib/python3.10/site-packages (from torchvision->sentence-transformers==2.2.2) (9.5.0)\r\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /opt/conda/lib/python3.10/site-packages (from packaging>=20.9->huggingface-hub>=0.4.0->sentence-transformers==2.2.2) (3.0.9)\r\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /opt/conda/lib/python3.10/site-packages (from jinja2->torch>=1.6.0->sentence-transformers==2.2.2) (2.1.3)\r\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /opt/conda/lib/python3.10/site-packages (from requests->huggingface-hub>=0.4.0->sentence-transformers==2.2.2) (3.1.0)\r\n",
      "Requirement already satisfied: idna<4,>=2.5 in /opt/conda/lib/python3.10/site-packages (from requests->huggingface-hub>=0.4.0->sentence-transformers==2.2.2) (3.4)\r\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /opt/conda/lib/python3.10/site-packages (from requests->huggingface-hub>=0.4.0->sentence-transformers==2.2.2) (1.26.15)\r\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.10/site-packages (from requests->huggingface-hub>=0.4.0->sentence-transformers==2.2.2) (2023.5.7)\r\n",
      "Requirement already satisfied: mpmath>=0.19 in /opt/conda/lib/python3.10/site-packages (from sympy->torch>=1.6.0->sentence-transformers==2.2.2) (1.3.0)\r\n",
      "Building wheels for collected packages: sentence-transformers\r\n",
      "  Building wheel for sentence-transformers (setup.py) ... \u001b[?25l-\b \b\\\b \b|\b \b/\b \bdone\r\n",
      "\u001b[?25h  Created wheel for sentence-transformers: filename=sentence_transformers-2.2.2-py3-none-any.whl size=125938 sha256=16e873074ef568e3b149003dd1304d52e366da97bd0c082ea14de65ff2a2ed82\r\n",
      "  Stored in directory: /root/.cache/pip/wheels/c6/26/e2/13ef17530724efc5be0bf3d290de1ecaa6c0ff0225fd548014\r\n",
      "Successfully built sentence-transformers\r\n",
      "Installing collected packages: sentence-transformers\r\n",
      "Successfully installed sentence-transformers-2.2.2\r\n"
     ]
    }
   ],
   "source": [
    "!pip install /kaggle/input/sci-llm-pip-v2/sentence-transformers-2.2.2.tar.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bd528554",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:02:40.318025Z",
     "iopub.status.busy": "2023-10-10T12:02:40.317708Z",
     "iopub.status.idle": "2023-10-10T12:02:40.372692Z",
     "shell.execute_reply": "2023-10-10T12:02:40.371724Z"
    },
    "papermill": {
     "duration": 0.065323,
     "end_time": "2023-10-10T12:02:40.374771",
     "exception": false,
     "start_time": "2023-10-10T12:02:40.309448",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prompt</th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>answer</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>What is the method of transcription in the lif...</td>\n",
       "      <td>RNA-templated transcription is the method of t...</td>\n",
       "      <td>Transcription occurs through a unique mechanis...</td>\n",
       "      <td>Reverse transcription is the method of transcr...</td>\n",
       "      <td>DNA-templated transcription is the method of t...</td>\n",
       "      <td>Transcription does not occur in the life cycle...</td>\n",
       "      <td>D</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>What is the role of the viral fiber glycoprote...</td>\n",
       "      <td>The viral fiber glycoproteins are involved in ...</td>\n",
       "      <td>The viral fiber glycoproteins code for 40 prot...</td>\n",
       "      <td>The viral fiber glycoproteins are responsible ...</td>\n",
       "      <td>The viral fiber glycoproteins mediate endocyto...</td>\n",
       "      <td>The viral fiber glycoproteins are responsible ...</td>\n",
       "      <td>D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>What is the significance of the faint Hα emiss...</td>\n",
       "      <td>The emission lines indicate that 3 Geminorum i...</td>\n",
       "      <td>The emission lines indicate that 3 Geminorum i...</td>\n",
       "      <td>The emission lines indicate that 3 Geminorum i...</td>\n",
       "      <td>The emission lines indicate that 3 Geminorum i...</td>\n",
       "      <td>The emission lines indicate that 3 Geminorum i...</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>What is the significance of the pedicellariae ...</td>\n",
       "      <td>They are used for climbing on corals.</td>\n",
       "      <td>They resemble the traps of the Venus fly trap ...</td>\n",
       "      <td>They are covered by short and stout spines.</td>\n",
       "      <td>They are found on the central disc of the sea ...</td>\n",
       "      <td>They are a characteristic feature of the Gonia...</td>\n",
       "      <td>B</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>What is the role of the microprocessor complex...</td>\n",
       "      <td>The microprocessor complex is responsible for ...</td>\n",
       "      <td>The microprocessor complex is responsible for ...</td>\n",
       "      <td>The microprocessor complex is involved in the ...</td>\n",
       "      <td>The microprocessor complex is involved in the ...</td>\n",
       "      <td>The microprocessor complex is responsible for ...</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              prompt  \\\n",
       "0  What is the method of transcription in the lif...   \n",
       "1  What is the role of the viral fiber glycoprote...   \n",
       "2  What is the significance of the faint Hα emiss...   \n",
       "3  What is the significance of the pedicellariae ...   \n",
       "4  What is the role of the microprocessor complex...   \n",
       "\n",
       "                                                   A  \\\n",
       "0  RNA-templated transcription is the method of t...   \n",
       "1  The viral fiber glycoproteins are involved in ...   \n",
       "2  The emission lines indicate that 3 Geminorum i...   \n",
       "3              They are used for climbing on corals.   \n",
       "4  The microprocessor complex is responsible for ...   \n",
       "\n",
       "                                                   B  \\\n",
       "0  Transcription occurs through a unique mechanis...   \n",
       "1  The viral fiber glycoproteins code for 40 prot...   \n",
       "2  The emission lines indicate that 3 Geminorum i...   \n",
       "3  They resemble the traps of the Venus fly trap ...   \n",
       "4  The microprocessor complex is responsible for ...   \n",
       "\n",
       "                                                   C  \\\n",
       "0  Reverse transcription is the method of transcr...   \n",
       "1  The viral fiber glycoproteins are responsible ...   \n",
       "2  The emission lines indicate that 3 Geminorum i...   \n",
       "3        They are covered by short and stout spines.   \n",
       "4  The microprocessor complex is involved in the ...   \n",
       "\n",
       "                                                   D  \\\n",
       "0  DNA-templated transcription is the method of t...   \n",
       "1  The viral fiber glycoproteins mediate endocyto...   \n",
       "2  The emission lines indicate that 3 Geminorum i...   \n",
       "3  They are found on the central disc of the sea ...   \n",
       "4  The microprocessor complex is involved in the ...   \n",
       "\n",
       "                                                   E answer  id  \n",
       "0  Transcription does not occur in the life cycle...      D   0  \n",
       "1  The viral fiber glycoproteins are responsible ...      D   1  \n",
       "2  The emission lines indicate that 3 Geminorum i...      A   2  \n",
       "3  They are a characteristic feature of the Gonia...      B   3  \n",
       "4  The microprocessor complex is responsible for ...      A   4  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "data_path = Path('/kaggle/input/kaggle-llm-science-exam')\n",
    "\n",
    "\n",
    "if os.getenv('KAGGLE_IS_COMPETITION_RERUN'):\n",
    "    test = pd.read_csv(data_path / 'test.csv')\n",
    "    CALC_SCORE = False\n",
    "else:\n",
    "    test = pd.read_csv(\"/kaggle/input/dataset-wiki-new-1/dataset_wiki_new_1_balanced.csv\")\n",
    "    test[\"id\"] = np.arange(len(test))\n",
    "#     test = pd.read_csv(data_path / 'train.csv')\n",
    "    CALC_SCORE = True\n",
    "test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "87ffe4b5",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:02:40.391488Z",
     "iopub.status.busy": "2023-10-10T12:02:40.391208Z",
     "iopub.status.idle": "2023-10-10T12:02:40.467559Z",
     "shell.execute_reply": "2023-10-10T12:02:40.466625Z"
    },
    "papermill": {
     "duration": 0.086918,
     "end_time": "2023-10-10T12:02:40.469605",
     "exception": false,
     "start_time": "2023-10-10T12:02:40.382687",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "test.to_parquet(\"test_raw.pq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c3a53ad4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:02:40.486450Z",
     "iopub.status.busy": "2023-10-10T12:02:40.485946Z",
     "iopub.status.idle": "2023-10-10T12:02:40.499765Z",
     "shell.execute_reply": "2023-10-10T12:02:40.498297Z"
    },
    "papermill": {
     "duration": 0.024423,
     "end_time": "2023-10-10T12:02:40.501673",
     "exception": false,
     "start_time": "2023-10-10T12:02:40.477250",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing get_topk.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile get_topk.py\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "from glob import glob\n",
    "import torch\n",
    "import numpy as np\n",
    "from sentence_transformers import SentenceTransformer\n",
    "from tqdm import tqdm\n",
    "from pathlib import Path\n",
    "from joblib import Parallel, delayed\n",
    "import argparse\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    \n",
    "    ap = argparse.ArgumentParser()\n",
    "    ap.add_argument('--wiki', type=str, required=True)\n",
    "    ap.add_argument('--model_name', type=str, required=True)\n",
    "    ap.add_argument('--test_file', type=str, required=True)\n",
    "    ap.add_argument('--topk', type=int, required=True)\n",
    "    ap.add_argument('--ind', type=int, required=True)\n",
    "    args = ap.parse_args()\n",
    "\n",
    "    if args.ind == 1:\n",
    "        TOP_K = 5 * 5\n",
    "    else:\n",
    "        if args.topk == 10:\n",
    "            TOP_K = 20\n",
    "        else:\n",
    "            TOP_K = 10\n",
    "\n",
    "    data_path = Path('/kaggle/input/kaggle-llm-science-exam')\n",
    "    \n",
    "    if args.wiki == \"cirrus\":\n",
    "        print(\"Using cirrus wiki\")\n",
    "        files_all = sorted(list(glob(\"/kaggle/input/cirruswiki-titles/*.parquet\")))\n",
    "    elif args.wiki == \"new\":\n",
    "        print(\"Using new wiki\")\n",
    "        files_all = sorted(list(glob(\"/kaggle/input/newwiki-titles/*.parquet\")))\n",
    "    elif args.wiki == \"last\":\n",
    "        print(\"Using last wiki\")\n",
    "        files_all = sorted(list(glob(\"/kaggle/input/cirruswiki-256-titles/*.parquet\"))) \n",
    "        \n",
    "    if args.wiki == \"cirrus\":\n",
    "        if \"e5-base\" in args.model_name:\n",
    "            files_np = sorted(list(glob(\"/kaggle/input/cirruswiki-e5-part?/*.npy\")))\n",
    "            files_np += sorted(list(glob(\"/kaggle/input/cirruswiki-e5-part9-1/*.npy\")))\n",
    "        elif \"e5-large\" in args.model_name:\n",
    "            files_np = sorted(list(glob(\"/kaggle/input/enwiki-cirrus-20230701-e5-large-part*/*.npy\")))\n",
    "        elif \"bge-base-15\" in args.model_name:\n",
    "            files_np = sorted(list(glob(\"/kaggle/input/cirruswiki512-bge-base-en-v1-5-part*/*.npy\")))\n",
    "        elif \"gte-base\" in args.model_name:\n",
    "            files_np = sorted(list(glob(\"/kaggle/input/cirruswiki512-gte-base-part*/*.npy\")))\n",
    "        elif \"bge-large\" in args.model_name:\n",
    "            files_np = sorted(list(glob(\"/kaggle/input/enwiki-cirrus-20230701-bge-large-part*/*.npy\")))\n",
    "        elif \"gte-large\" in args.model_name:\n",
    "            files_np = sorted(list(glob(\"/kaggle/input/enwiki-cirrus-20230701-gte-large-part*/*.npy\")))\n",
    "    elif args.wiki == \"new\":\n",
    "        if \"gte-base-title\" in args.model_name:\n",
    "            files_np = sorted(list(glob(\"/kaggle/input/gte-base-title-part*/*.npy\")))\n",
    "        elif \"e5-base-title\" in args.model_name:\n",
    "            files_np = sorted(list(glob(\"/kaggle/input/newwiki-e5-base-title-part*/*.npy\")))\n",
    "        elif \"e5-base\" in args.model_name:\n",
    "            files_np = sorted(list(glob(\"/kaggle/input/newwiki-e5-base-part*/*.npy\")))\n",
    "        elif \"gte-base\" in args.model_name:\n",
    "            files_np = sorted(list(glob(\"/kaggle/input/newwiki-gte-part*/*.npy\")))\n",
    "        elif \"gte-large\" in args.model_name:\n",
    "            files_np = sorted(list(glob(\"/kaggle/input/wiki31m-gte-large-title-p*/*.npy\")))\n",
    "        elif \"bge-base\" in args.model_name:\n",
    "            files_np = sorted(list(glob(\"/kaggle/input/wiki31m-bge-base-title-part*/*.npy\")))\n",
    "        elif \"e5-large\" in args.model_name:\n",
    "            files_np = sorted(list(glob(\"/kaggle/input/wiki-30-08-e5-large-title-p*/*.npy\")))\n",
    "        elif \"bge-large\" in args.model_name:\n",
    "            files_np = sorted(list(glob(\"/kaggle/input/wiki31m-bge-large-title-p*/*.npy\")))\n",
    "    elif args.wiki == \"last\":\n",
    "        files_np = sorted(list(glob(\"/kaggle/input/enwiki-cirrus-256-e5-base-part*/*.npy\")))       \n",
    "\n",
    "    files_all = [(x, y) for x, y in zip(files_all, files_np)]\n",
    "    files = [files_all[:len(files_all)//2], files_all[len(files_all)//2:]]\n",
    "\n",
    "\n",
    "    if \"e5-base\" in args.model_name:\n",
    "        model = SentenceTransformer('/kaggle/input/intfloat-e5-base-v2').to(\"cuda:0\")\n",
    "    elif \"e5-large\" in args.model_name:\n",
    "        model = SentenceTransformer('/kaggle/input/intfloat-e5-large-v2').to(\"cuda:0\")\n",
    "    elif \"bge-base-15\" in args.model_name: \n",
    "        model = SentenceTransformer('/kaggle/input/baai-bge-base-en-v1-5').to(\"cuda:0\")\n",
    "    elif \"gte-base\" in args.model_name: \n",
    "        model = SentenceTransformer('/kaggle/input/thenlper-gte-base').to(\"cuda:0\")\n",
    "    elif \"gte-large\" in args.model_name:\n",
    "        model = SentenceTransformer('/kaggle/input/thenlper-gte-large').to(\"cuda:0\")\n",
    "    elif \"bge-base\" in args.model_name:\n",
    "        model = SentenceTransformer('/kaggle/input/baai-bge-base-en').to(\"cuda:0\")\n",
    "    elif \"bge-large\" in args.model_name:\n",
    "        model = SentenceTransformer('/kaggle/input/baai-bge-large-en').to(\"cuda:0\")\n",
    "\n",
    "    test = pd.read_parquet(\"test_raw.pq\")\n",
    "\n",
    "    embs = []\n",
    "    for idx, row in tqdm(test.iterrows(), total=len(test)):\n",
    "        if \"e5\" in args.model_name:\n",
    "            if args.ind == 1:\n",
    "                sentences = [\n",
    "                    \"query: \" + row.prompt + \" \" + row.A,\n",
    "                    \"query: \" + row.prompt + \" \" + row.B,\n",
    "                    \"query: \" + row.prompt + \" \" + row.C,\n",
    "                    \"query: \" + row.prompt + \" \" + row.D,\n",
    "                    \"query: \" + row.prompt + \" \" + row.E,\n",
    "                ]\n",
    "            else:\n",
    "                sentences = [\"query: \" + row.prompt + \" \" + row.A + \" \" + row.B + \" \" + row.C + \" \" + row.D + \" \" + row.E]\n",
    "        elif \"bge\" in args.model_name:\n",
    "            if args.ind == 1:\n",
    "                sentences = [\n",
    "                    \"Represent this sentence for searching relevant passages: \" + row.prompt + \" \" + row.A,\n",
    "                    \"Represent this sentence for searching relevant passages: \" + row.prompt + \" \" + row.B,\n",
    "                    \"Represent this sentence for searching relevant passages: \" + row.prompt + \" \" + row.C,\n",
    "                    \"Represent this sentence for searching relevant passages: \" + row.prompt + \" \" + row.D,\n",
    "                    \"Represent this sentence for searching relevant passages: \" + row.prompt + \" \" + row.E,\n",
    "                ]\n",
    "            else:\n",
    "                sentences = [\"Represent this sentence for searching relevant passages: \" + row.prompt + \" \" + row.A + \" \" + row.B + \" \" + row.C + \" \" + row.D + \" \" + row.E]    \n",
    "        elif \"gte\" in args.model_name:\n",
    "            if args.ind == 1:\n",
    "                sentences = [\n",
    "                    row.prompt + \" \" + row.A,\n",
    "                    row.prompt + \" \" + row.B,\n",
    "                    row.prompt + \" \" + row.C,\n",
    "                    row.prompt + \" \" + row.D,\n",
    "                    row.prompt + \" \" + row.E,\n",
    "                ]\n",
    "            else:\n",
    "                sentences = [row.prompt + \" \" + row.A + \" \" + row.B + \" \" + row.C + \" \" + row.D + \" \" + row.E]\n",
    "\n",
    "        embeddings = torch.Tensor(model.encode(sentences, show_progress_bar=False, normalize_embeddings=True))\n",
    "        embs.append(torch.nn.functional.normalize(embeddings, dim=1))\n",
    "\n",
    "    if args.ind == 1:\n",
    "        query_embeddings = torch.Tensor(np.concatenate(embs)).squeeze(1)\n",
    "    else:\n",
    "        query_embeddings = torch.Tensor(np.stack(embs)).squeeze(1)\n",
    "    print(query_embeddings.shape)\n",
    "\n",
    "    # Create placeholders for top-k matches\n",
    "    all_vals_gpu_0 = torch.full((len(test), TOP_K), -float('inf'), dtype=torch.float16)\n",
    "    all_texts_gpu_0 = [[None] * TOP_K for _ in range(len(all_vals_gpu_0))]\n",
    "\n",
    "    all_vals_gpu_1 = torch.full((len(test), TOP_K), -float('inf'), dtype=torch.float16)\n",
    "    all_texts_gpu_1 = [[None] * TOP_K for _ in range(len(all_vals_gpu_1))]\n",
    "\n",
    "    print(all_vals_gpu_0.shape)\n",
    "    print(len(all_texts_gpu_0))\n",
    "\n",
    "\n",
    "    def cos_similarity_matrix(a: torch.Tensor, b: torch.Tensor, eps: float = 1e-8):\n",
    "        \"\"\"Calculates cosine similarities between tensor a and b.\"\"\"\n",
    "        sim_mt = torch.mm(a, b.transpose(0, 1))\n",
    "        return sim_mt\n",
    "\n",
    "\n",
    "    def get_topk(embeddings_from, embeddings_to, topk=1000, bs=512):\n",
    "        chunk = bs\n",
    "        embeddings_chunks = embeddings_from.split(chunk)\n",
    "\n",
    "        vals = []\n",
    "        inds = []\n",
    "        for idx in tqdm(range(len(embeddings_chunks))):\n",
    "            cos_sim_chunk = cos_similarity_matrix(embeddings_chunks[idx].to(embeddings_to.device).half(), embeddings_to).float()\n",
    "\n",
    "            cos_sim_chunk = torch.nan_to_num(cos_sim_chunk, nan=0.0)\n",
    "\n",
    "            topk = min(topk, cos_sim_chunk.size(1))\n",
    "            vals_chunk, inds_chunk = torch.topk(cos_sim_chunk, k=topk, dim=1)\n",
    "            vals.append(vals_chunk[:, :].detach().cpu())\n",
    "            inds.append(inds_chunk[:, :].detach().cpu())\n",
    "\n",
    "            del vals_chunk\n",
    "            del inds_chunk\n",
    "            del cos_sim_chunk\n",
    "\n",
    "        vals = torch.cat(vals).detach().cpu()\n",
    "        inds = torch.cat(inds).detach().cpu()\n",
    "\n",
    "        return inds, vals\n",
    "\n",
    "\n",
    "    def insert_value_at(tensor, value, position):\n",
    "        # Ensure the position is valid\n",
    "        if position < 0 or position >= len(tensor):\n",
    "            raise ValueError(\"Position should be between 0 and tensor length - 1.\")\n",
    "\n",
    "        # Slice the tensor into two parts\n",
    "        left = tensor[:position]\n",
    "        right = tensor[position:]\n",
    "\n",
    "        # Create a tensor for the value to be inserted\n",
    "        value_tensor = torch.tensor([value], dtype=tensor.dtype)\n",
    "\n",
    "        # Concatenate the tensors together and slice to the original length\n",
    "        result = torch.cat([left, value_tensor, right])[:-1]\n",
    "\n",
    "        return result\n",
    "\n",
    "\n",
    "    def insert_value_at_list(lst, value, position):\n",
    "        # Ensure the position is valid\n",
    "        if position < 0 or position >= len(lst):\n",
    "            raise ValueError(\"Position should be between 0 and list length - 1.\")\n",
    "\n",
    "        # Insert value at the specified position\n",
    "        lst.insert(position, value)\n",
    "\n",
    "        # Remove the last value to maintain original length\n",
    "        lst.pop()\n",
    "\n",
    "        return lst\n",
    "    \n",
    "    from collections import OrderedDict\n",
    "\n",
    "    def get_first_5_unique_strings(input_list):\n",
    "#         print(len(input_list))\n",
    "        unique_strings = list(OrderedDict.fromkeys(input_list))\n",
    "#         print(unique_strings)\n",
    "        while len(unique_strings) < args.topk:\n",
    "            # TODO: check why this is ever happening\n",
    "            unique_strings.append(\"No context found\")\n",
    "            print(\"appending None\")\n",
    "        return unique_strings[:args.topk]\n",
    "\n",
    "\n",
    "    def load_data(files, device):\n",
    "        for file, file_np in tqdm(files):\n",
    "            df = pd.read_parquet(file, engine=\"pyarrow\", use_threads=True)\n",
    "            file_embeddings = np.load(file_np)\n",
    "\n",
    "            data_embeddings = torch.Tensor(file_embeddings).to(device).half()\n",
    "            data_embeddings = torch.nn.functional.normalize(data_embeddings, dim=1)\n",
    "\n",
    "            max_inds, max_vals = get_topk(query_embeddings, data_embeddings, topk=TOP_K, bs=8)\n",
    "\n",
    "            # loop through all queries (test)\n",
    "            for i in range(len(test)):\n",
    "                if args.ind == 1:\n",
    "                    for answer in range(5):\n",
    "                        for new in range(TOP_K):\n",
    "                            if device == \"cuda:0\":\n",
    "                                if max_vals[i * 5 + answer][new].item() < all_vals_gpu_0[i][TOP_K - 1]:\n",
    "                                    break\n",
    "                                for old in range(TOP_K):\n",
    "                                    if max_vals[i * 5 + answer][new].item() > all_vals_gpu_0[i][old]:\n",
    "                                        all_vals_gpu_0[i] = insert_value_at(all_vals_gpu_0[i], value=max_vals[i * 5 + answer][new].item(), position=old)\n",
    "                                        all_texts_gpu_0[i] = insert_value_at_list(all_texts_gpu_0[i], value=df.iloc[max_inds[i * 5 + answer][new].item()].text, position=old)\n",
    "                                        break\n",
    "                            else:\n",
    "                                if max_vals[i * 5 + answer][new].item() < all_vals_gpu_1[i][TOP_K - 1]:\n",
    "                                    break\n",
    "                                for old in range(TOP_K):\n",
    "                                    if max_vals[i * 5 + answer][new].item() > all_vals_gpu_1[i][old]:\n",
    "                                        all_vals_gpu_1[i] = insert_value_at(all_vals_gpu_1[i], value=max_vals[i * 5 + answer][new].item(), position=old)\n",
    "                                        all_texts_gpu_1[i] = insert_value_at_list(all_texts_gpu_1[i], value=df.iloc[max_inds[i * 5 + answer][new].item()].text, position=old)\n",
    "                                        break\n",
    "                else:\n",
    "                    # start with highest new val (pos 0) vs worst value already in the toplist (pos topk - 1)\n",
    "                    for new in range(TOP_K):\n",
    "                        if device == \"cuda:0\":\n",
    "                            if max_vals[i][new].item() < all_vals_gpu_0[i][TOP_K - 1]:\n",
    "                                break\n",
    "                            for old in range(TOP_K):\n",
    "                                if max_vals[i][new].item() > all_vals_gpu_0[i][old]:\n",
    "                                    all_vals_gpu_0[i] = insert_value_at(all_vals_gpu_0[i], value=max_vals[i][new].item(), position=old)\n",
    "                                    all_texts_gpu_0[i] = insert_value_at_list(all_texts_gpu_0[i], value=df.iloc[max_inds[i][new].item()].text, position=old)\n",
    "                                    break\n",
    "                        else:\n",
    "                            if max_vals[i][new].item() < all_vals_gpu_1[i][TOP_K - 1]:\n",
    "                                break\n",
    "                            for old in range(TOP_K):\n",
    "                                if max_vals[i][new].item() > all_vals_gpu_1[i][old]:\n",
    "                                    all_vals_gpu_1[i] = insert_value_at(all_vals_gpu_1[i], value=max_vals[i][new].item(), position=old)\n",
    "                                    all_texts_gpu_1[i] = insert_value_at_list(all_texts_gpu_1[i], value=df.iloc[max_inds[i][new].item()].text, position=old)\n",
    "                                    break\n",
    "\n",
    "            if not os.getenv('KAGGLE_IS_COMPETITION_RERUN'):\n",
    "                break\n",
    "\n",
    "\n",
    "    Parallel(n_jobs=2, backend=\"threading\")(delayed(load_data)(files[i], f\"cuda:{i}\") for i in range(2))\n",
    "    all_vals = torch.hstack([all_vals_gpu_0, all_vals_gpu_1])\n",
    "    val, inds = torch.topk(all_vals.float(), axis=1, k=TOP_K)\n",
    "    all_texts = [[(t0 + t1)[inner_idx.item()] for inner_idx in idx] for t0, t1, idx in zip(all_texts_gpu_0, all_texts_gpu_1, inds)]\n",
    "    # all_vals = [[(list(t0) + list(t1))[inner_idx.item()] for inner_idx in idx] for t0, t1, idx in zip(all_vals_gpu_0, all_vals_gpu_1, inds)]\n",
    "\n",
    "\n",
    "    def remove_consecutive_duplicates(input_list):\n",
    "        if not input_list:\n",
    "            return [' '] * args.topk\n",
    "\n",
    "        new_list = [input_list[0]]\n",
    "        for i in range(1, len(input_list)):\n",
    "            if input_list[i] != input_list[i-1]:\n",
    "                new_list.append(input_list[i])\n",
    "\n",
    "        # Append empty strings if new_list length is less than 5\n",
    "        while len(new_list) < args.topk:\n",
    "            new_list.append(' ')\n",
    "\n",
    "        return new_list\n",
    "    \n",
    "    if args.ind == 1:\n",
    "        all_texts = [get_first_5_unique_strings(x) for x in all_texts]\n",
    "    else:\n",
    "        all_texts = [remove_consecutive_duplicates(lst) for lst in all_texts]\n",
    "\n",
    "    test[\"context\"] = [\"\\n###\\n\".join([x[i] for i in list(range(args.topk))[::-1]]) for x in all_texts]\n",
    "\n",
    "    test[\"context_v2\"] = [\"Context 4: \" + x[4] + \"\\n###\\n\" + \"Context 3: \" + x[3] + \"\\n###\\n\" + \"Context 2: \" + x[2] + \"\\n###\\n\" + \"Context 1: \" + x[1] + \"\\n###\\n\" + \"Context 0: \" + x[0] for x in all_texts]\n",
    "\n",
    "    test[\"context_deberta\"] = [\" [SEP] \".join(x[:args.topk]) for x in all_texts]\n",
    "\n",
    "    print(test[\"context\"].values[0])\n",
    "    print(\"\\nV2:\\n\")\n",
    "    print(test[\"context_v2\"].values[0])\n",
    "    print(\"\\nDEBERTA:\\n\")\n",
    "    print(test[\"context_deberta\"].values[0])\n",
    "\n",
    "    test.to_parquet(f\"{args.test_file}.pq\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bf1f86d8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:02:40.517864Z",
     "iopub.status.busy": "2023-10-10T12:02:40.517233Z",
     "iopub.status.idle": "2023-10-10T12:04:13.647339Z",
     "shell.execute_reply": "2023-10-10T12:04:13.646176Z"
    },
    "papermill": {
     "duration": 93.140721,
     "end_time": "2023-10-10T12:04:13.649669",
     "exception": false,
     "start_time": "2023-10-10T12:02:40.508948",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5\r\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\r\n",
      "Using cirrus wiki\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']\r\n",
      "  warnings.warn(f\"unable to load libtensorflow_io_plugins.so: {e}\")\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tensorflow13GcsFileSystemE']\r\n",
      "  warnings.warn(f\"file system plugins are not loaded: {e}\")\r\n",
      "100%|█████████████████████████████████████████| 300/300 [00:16<00:00, 18.61it/s]\r\n",
      "torch.Size([300, 1024])\r\n",
      "torch.Size([300, 10])\r\n",
      "300\r\n",
      "  0%|                                                   | 0/117 [00:00<?, ?it/s]\r\n",
      "  0%|                                                   | 0/118 [00:00<?, ?it/s]\u001b[A\r\n",
      "\r\n",
      "  0%|                                                    | 0/38 [00:00<?, ?it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      "100%|██████████████████████████████████████████| 38/38 [00:00<00:00, 218.28it/s]\r\n",
      "  0%|                                                   | 0/117 [00:12<?, ?it/s]\r\n",
      "100%|███████████████████████████████████████████| 38/38 [00:00<00:00, 57.87it/s]\r\n",
      "  0%|                                                   | 0/118 [00:15<?, ?it/s]\r\n",
      "See Herpesvirales#Taxonomy for information on taxonomic history, phylogenetic research, and the nomenclatural system. All members of the Herpesviridae share a common structure; a relatively large, monopartite, double-stranded, linear DNA genome encoding 100-200 genes encased within an icosahedral protein cage (with T=16 symmetry) called the capsid, which is itself wrapped in a protein layer called the tegument containing both viral proteins and viral mRNAs and a lipid bilayer membrane called the envelope.\r\n",
      "###\r\n",
      "Thogotovirus is a genus of enveloped RNA viruses, one of seven genera in the virus family Orthomyxoviridae. Their single-stranded, negative-sense RNA genome has six or seven segments. Thogotoviruses are distinguished from most other orthomyxoviruses by being arboviruses – viruses that are transmitted by arthropods, in this case usually ticks. Thogotoviruses can replicate in both tick cells and vertebrate cells; one subtype has also been isolated from mosquitoes.\r\n",
      "###\r\n",
      "Some notable pathogenic viruses cause smallpox, influenza, mumps, measles, chickenpox, ebola, and rubella. Viruses typically range between 20 and 300 nanometers in length. This type of pathogen is not cellular, and is instead composed of either RNA (Ribonucleic acid) or DNA (Deoxyribonucleic acid) within a protein shell - the capsid. Pathogenic viruses infiltrate host cells and manipulate the organelles within the cell such as the Ribosomes, Golgi Apparatus, and Endoplasmic Reticulum in order to multiply which commonly results in the death of the host cell via cellular decay.\r\n",
      "###\r\n",
      "The icosahedral nucleocapsid is similar to that of tailed bacteriophage in the order Caudovirales. This capsid has 161 capsomers consisting of 150 hexons and 11 pentons, as well as a portal complex that allows entry and exit of DNA into the capsid. All herpesviruses are nuclear-replicating—the viral DNA is transcribed to mRNA within the infected cell's nucleus.[citation needed] Infection is initiated when a viral particle contacts a cell with specific types of receptor molecules on the cell surface.\r\n",
      "###\r\n",
      "Viral replication is cytoplasmic. Entry into the host cell is achieved by attachment of the viral envelope protein E to host receptors, which mediates clathrin-mediated endocytosis. Replication follows the positive-stranded RNA virus replication model. Positive-stranded RNA virus transcription is the method of transcription. Translation takes place by viral initiation. The virus exits the host cell by budding. Humans and mammals serve as the natural hosts. The virus is transmitted via vectors (ticks and mosquitoes).\r\n",
      "\r\n",
      "V2:\r\n",
      "\r\n",
      "Context 4: See Herpesvirales#Taxonomy for information on taxonomic history, phylogenetic research, and the nomenclatural system. All members of the Herpesviridae share a common structure; a relatively large, monopartite, double-stranded, linear DNA genome encoding 100-200 genes encased within an icosahedral protein cage (with T=16 symmetry) called the capsid, which is itself wrapped in a protein layer called the tegument containing both viral proteins and viral mRNAs and a lipid bilayer membrane called the envelope.\r\n",
      "###\r\n",
      "Context 3: Thogotovirus is a genus of enveloped RNA viruses, one of seven genera in the virus family Orthomyxoviridae. Their single-stranded, negative-sense RNA genome has six or seven segments. Thogotoviruses are distinguished from most other orthomyxoviruses by being arboviruses – viruses that are transmitted by arthropods, in this case usually ticks. Thogotoviruses can replicate in both tick cells and vertebrate cells; one subtype has also been isolated from mosquitoes.\r\n",
      "###\r\n",
      "Context 2: Some notable pathogenic viruses cause smallpox, influenza, mumps, measles, chickenpox, ebola, and rubella. Viruses typically range between 20 and 300 nanometers in length. This type of pathogen is not cellular, and is instead composed of either RNA (Ribonucleic acid) or DNA (Deoxyribonucleic acid) within a protein shell - the capsid. Pathogenic viruses infiltrate host cells and manipulate the organelles within the cell such as the Ribosomes, Golgi Apparatus, and Endoplasmic Reticulum in order to multiply which commonly results in the death of the host cell via cellular decay.\r\n",
      "###\r\n",
      "Context 1: The icosahedral nucleocapsid is similar to that of tailed bacteriophage in the order Caudovirales. This capsid has 161 capsomers consisting of 150 hexons and 11 pentons, as well as a portal complex that allows entry and exit of DNA into the capsid. All herpesviruses are nuclear-replicating—the viral DNA is transcribed to mRNA within the infected cell's nucleus.[citation needed] Infection is initiated when a viral particle contacts a cell with specific types of receptor molecules on the cell surface.\r\n",
      "###\r\n",
      "Context 0: Viral replication is cytoplasmic. Entry into the host cell is achieved by attachment of the viral envelope protein E to host receptors, which mediates clathrin-mediated endocytosis. Replication follows the positive-stranded RNA virus replication model. Positive-stranded RNA virus transcription is the method of transcription. Translation takes place by viral initiation. The virus exits the host cell by budding. Humans and mammals serve as the natural hosts. The virus is transmitted via vectors (ticks and mosquitoes).\r\n",
      "\r\n",
      "DEBERTA:\r\n",
      "\r\n",
      "Viral replication is cytoplasmic. Entry into the host cell is achieved by attachment of the viral envelope protein E to host receptors, which mediates clathrin-mediated endocytosis. Replication follows the positive-stranded RNA virus replication model. Positive-stranded RNA virus transcription is the method of transcription. Translation takes place by viral initiation. The virus exits the host cell by budding. Humans and mammals serve as the natural hosts. The virus is transmitted via vectors (ticks and mosquitoes). [SEP] The icosahedral nucleocapsid is similar to that of tailed bacteriophage in the order Caudovirales. This capsid has 161 capsomers consisting of 150 hexons and 11 pentons, as well as a portal complex that allows entry and exit of DNA into the capsid. All herpesviruses are nuclear-replicating—the viral DNA is transcribed to mRNA within the infected cell's nucleus.[citation needed] Infection is initiated when a viral particle contacts a cell with specific types of receptor molecules on the cell surface. [SEP] Some notable pathogenic viruses cause smallpox, influenza, mumps, measles, chickenpox, ebola, and rubella. Viruses typically range between 20 and 300 nanometers in length. This type of pathogen is not cellular, and is instead composed of either RNA (Ribonucleic acid) or DNA (Deoxyribonucleic acid) within a protein shell - the capsid. Pathogenic viruses infiltrate host cells and manipulate the organelles within the cell such as the Ribosomes, Golgi Apparatus, and Endoplasmic Reticulum in order to multiply which commonly results in the death of the host cell via cellular decay. [SEP] Thogotovirus is a genus of enveloped RNA viruses, one of seven genera in the virus family Orthomyxoviridae. Their single-stranded, negative-sense RNA genome has six or seven segments. Thogotoviruses are distinguished from most other orthomyxoviruses by being arboviruses – viruses that are transmitted by arthropods, in this case usually ticks. Thogotoviruses can replicate in both tick cells and vertebrate cells; one subtype has also been isolated from mosquitoes. [SEP] See Herpesvirales#Taxonomy for information on taxonomic history, phylogenetic research, and the nomenclatural system. All members of the Herpesviridae share a common structure; a relatively large, monopartite, double-stranded, linear DNA genome encoding 100-200 genes encased within an icosahedral protein cage (with T=16 symmetry) called the capsid, which is itself wrapped in a protein layer called the tegument containing both viral proteins and viral mRNAs and a lipid bilayer membrane called the envelope.\r\n"
     ]
    }
   ],
   "source": [
    "!python get_topk.py --wiki \"cirrus\" --model_name \"e5-large\" --test_file \"test\" --topk 5 --ind 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c02c2899",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:04:13.677600Z",
     "iopub.status.busy": "2023-10-10T12:04:13.677312Z",
     "iopub.status.idle": "2023-10-10T12:05:00.040999Z",
     "shell.execute_reply": "2023-10-10T12:05:00.039843Z"
    },
    "papermill": {
     "duration": 46.380319,
     "end_time": "2023-10-10T12:05:00.043157",
     "exception": false,
     "start_time": "2023-10-10T12:04:13.662838",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5\r\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\r\n",
      "Using new wiki\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']\r\n",
      "  warnings.warn(f\"unable to load libtensorflow_io_plugins.so: {e}\")\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tensorflow13GcsFileSystemE']\r\n",
      "  warnings.warn(f\"file system plugins are not loaded: {e}\")\r\n",
      "100%|█████████████████████████████████████████| 300/300 [00:05<00:00, 55.10it/s]\r\n",
      "torch.Size([300, 768])\r\n",
      "torch.Size([300, 10])\r\n",
      "300\r\n",
      "  0%|                                                    | 0/62 [00:00<?, ?it/s]\r\n",
      "  0%|                                                    | 0/62 [00:00<?, ?it/s]\u001b[A\r\n",
      "\r\n",
      "  0%|                                                    | 0/38 [00:00<?, ?it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      "100%|██████████████████████████████████████████| 38/38 [00:00<00:00, 286.72it/s]\r\n",
      "  0%|                                                    | 0/62 [00:09<?, ?it/s]\r\n",
      "100%|███████████████████████████████████████████| 38/38 [00:00<00:00, 59.74it/s]\r\n",
      "  0%|                                                    | 0/62 [00:12<?, ?it/s]\r\n",
      "Viral replication is cytoplasmic. Entry into the host cell is achieved by penetration into the host cell. Replication follows the positive stranded RNA virus replication model. Positive stranded RNA virus transcription, using the premature termination model of subgenomic RNA transcription is the method of transcription. Translation takes place by -1 ribosomal frameshifting. The virus exits the host cell by tubule-guided viral movement. Plants serve as the natural host. Transmission routes are mechanical, seed borne, and contact.\r\n",
      "###\r\n",
      "The genus contains the following species: Viruses in \"Teschovirus\" are non-enveloped, with icosahedral, spherical, and round geometries, and T=pseudo3 symmetry. The diameter is around 30 nm. Viral replication is cytoplasmic. Entry into the host cell is achieved by attachment of the virus to host receptors, which mediates endocytosis. Replication follows the positive stranded RNA virus replication model. Positive stranded RNA virus transcription is the method of transcription. Translation takes place by ribosomal skipping. The virus exits the host cell by lysis, and viroporins. Pig serve as the natural host. Transmission routes are fecal-oral.\r\n",
      "###\r\n",
      "In the field of virology, reverse-genetics techniques can be used to recover full-length infectious viruses with desired mutations or insertions in the viral genomes or in specific virus genes. Technologies that allow these manipulations include circular polymerase extension reaction (CPER) which was first used to generate infectious cDNA for Kunjin virus a close relative of West Nile virus. CPER has also been successfully utilised to generate a range of positive-sense RNA viruses such as SARS-CoV-2, the causative agent of COVID-19.\r\n",
      "###\r\n",
      "Viral replication is cytoplasmic. Entry into the host cell is achieved by penetration into the host cell. Replication follows the positive stranded RNA virus replication model. Positive stranded RNA virus transcription is the method of transcription. Translation takes place by leaky scanning, -1 ribosomal frameshifting, and suppression of termination. The virus exits the host cell by tubule-guided viral movement. Plants serve as the natural host. The virus is transmitted via a vector (insects). Transmission routes are vector and mechanical.\r\n",
      "###\r\n",
      "\"Petuviruses\" (including other members of the \"Caulimovirusus\" such as \"soymoviruses\" and \"Cavemoviruses\") form isometric particles whereas members of \"Badnaviruses\" and \"Tungroviruses\" have bacilliform virus particles. Viral replication is nuclear/cytoplasmic. Entry into the host cell is achieved by attachment of the viral proteins to host receptors, which mediates endocytosis. Replication follows the dsDNA(RT) replication model. DNA-templated transcription, specifically dsDNA(RT) transcription is the method of transcription. The virus exits the host cell by nuclear pore export, and tubule-guided viral movement. Plants serve as the natural host. Transmission routes are mechanical and grafting.\r\n",
      "\r\n",
      "V2:\r\n",
      "\r\n",
      "Context 4: Viral replication is cytoplasmic. Entry into the host cell is achieved by penetration into the host cell. Replication follows the positive stranded RNA virus replication model. Positive stranded RNA virus transcription, using the premature termination model of subgenomic RNA transcription is the method of transcription. Translation takes place by -1 ribosomal frameshifting. The virus exits the host cell by tubule-guided viral movement. Plants serve as the natural host. Transmission routes are mechanical, seed borne, and contact.\r\n",
      "###\r\n",
      "Context 3: The genus contains the following species: Viruses in \"Teschovirus\" are non-enveloped, with icosahedral, spherical, and round geometries, and T=pseudo3 symmetry. The diameter is around 30 nm. Viral replication is cytoplasmic. Entry into the host cell is achieved by attachment of the virus to host receptors, which mediates endocytosis. Replication follows the positive stranded RNA virus replication model. Positive stranded RNA virus transcription is the method of transcription. Translation takes place by ribosomal skipping. The virus exits the host cell by lysis, and viroporins. Pig serve as the natural host. Transmission routes are fecal-oral.\r\n",
      "###\r\n",
      "Context 2: In the field of virology, reverse-genetics techniques can be used to recover full-length infectious viruses with desired mutations or insertions in the viral genomes or in specific virus genes. Technologies that allow these manipulations include circular polymerase extension reaction (CPER) which was first used to generate infectious cDNA for Kunjin virus a close relative of West Nile virus. CPER has also been successfully utilised to generate a range of positive-sense RNA viruses such as SARS-CoV-2, the causative agent of COVID-19.\r\n",
      "###\r\n",
      "Context 1: Viral replication is cytoplasmic. Entry into the host cell is achieved by penetration into the host cell. Replication follows the positive stranded RNA virus replication model. Positive stranded RNA virus transcription is the method of transcription. Translation takes place by leaky scanning, -1 ribosomal frameshifting, and suppression of termination. The virus exits the host cell by tubule-guided viral movement. Plants serve as the natural host. The virus is transmitted via a vector (insects). Transmission routes are vector and mechanical.\r\n",
      "###\r\n",
      "Context 0: \"Petuviruses\" (including other members of the \"Caulimovirusus\" such as \"soymoviruses\" and \"Cavemoviruses\") form isometric particles whereas members of \"Badnaviruses\" and \"Tungroviruses\" have bacilliform virus particles. Viral replication is nuclear/cytoplasmic. Entry into the host cell is achieved by attachment of the viral proteins to host receptors, which mediates endocytosis. Replication follows the dsDNA(RT) replication model. DNA-templated transcription, specifically dsDNA(RT) transcription is the method of transcription. The virus exits the host cell by nuclear pore export, and tubule-guided viral movement. Plants serve as the natural host. Transmission routes are mechanical and grafting.\r\n",
      "\r\n",
      "DEBERTA:\r\n",
      "\r\n",
      "\"Petuviruses\" (including other members of the \"Caulimovirusus\" such as \"soymoviruses\" and \"Cavemoviruses\") form isometric particles whereas members of \"Badnaviruses\" and \"Tungroviruses\" have bacilliform virus particles. Viral replication is nuclear/cytoplasmic. Entry into the host cell is achieved by attachment of the viral proteins to host receptors, which mediates endocytosis. Replication follows the dsDNA(RT) replication model. DNA-templated transcription, specifically dsDNA(RT) transcription is the method of transcription. The virus exits the host cell by nuclear pore export, and tubule-guided viral movement. Plants serve as the natural host. Transmission routes are mechanical and grafting. [SEP] Viral replication is cytoplasmic. Entry into the host cell is achieved by penetration into the host cell. Replication follows the positive stranded RNA virus replication model. Positive stranded RNA virus transcription is the method of transcription. Translation takes place by leaky scanning, -1 ribosomal frameshifting, and suppression of termination. The virus exits the host cell by tubule-guided viral movement. Plants serve as the natural host. The virus is transmitted via a vector (insects). Transmission routes are vector and mechanical. [SEP] In the field of virology, reverse-genetics techniques can be used to recover full-length infectious viruses with desired mutations or insertions in the viral genomes or in specific virus genes. Technologies that allow these manipulations include circular polymerase extension reaction (CPER) which was first used to generate infectious cDNA for Kunjin virus a close relative of West Nile virus. CPER has also been successfully utilised to generate a range of positive-sense RNA viruses such as SARS-CoV-2, the causative agent of COVID-19. [SEP] The genus contains the following species: Viruses in \"Teschovirus\" are non-enveloped, with icosahedral, spherical, and round geometries, and T=pseudo3 symmetry. The diameter is around 30 nm. Viral replication is cytoplasmic. Entry into the host cell is achieved by attachment of the virus to host receptors, which mediates endocytosis. Replication follows the positive stranded RNA virus replication model. Positive stranded RNA virus transcription is the method of transcription. Translation takes place by ribosomal skipping. The virus exits the host cell by lysis, and viroporins. Pig serve as the natural host. Transmission routes are fecal-oral. [SEP] Viral replication is cytoplasmic. Entry into the host cell is achieved by penetration into the host cell. Replication follows the positive stranded RNA virus replication model. Positive stranded RNA virus transcription, using the premature termination model of subgenomic RNA transcription is the method of transcription. Translation takes place by -1 ribosomal frameshifting. The virus exits the host cell by tubule-guided viral movement. Plants serve as the natural host. Transmission routes are mechanical, seed borne, and contact.\r\n"
     ]
    }
   ],
   "source": [
    "!python get_topk.py --wiki \"new\" --model_name \"e5-base-title\" --test_file \"test2\" --topk 5 --ind 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ef3e860f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:05:00.074944Z",
     "iopub.status.busy": "2023-10-10T12:05:00.074128Z",
     "iopub.status.idle": "2023-10-10T12:05:46.634142Z",
     "shell.execute_reply": "2023-10-10T12:05:46.632678Z"
    },
    "papermill": {
     "duration": 46.57858,
     "end_time": "2023-10-10T12:05:46.636580",
     "exception": false,
     "start_time": "2023-10-10T12:05:00.058000",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5\r\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\r\n",
      "Using new wiki\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']\r\n",
      "  warnings.warn(f\"unable to load libtensorflow_io_plugins.so: {e}\")\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tensorflow13GcsFileSystemE']\r\n",
      "  warnings.warn(f\"file system plugins are not loaded: {e}\")\r\n",
      "100%|█████████████████████████████████████████| 300/300 [00:12<00:00, 23.26it/s]\r\n",
      "torch.Size([300, 1024])\r\n",
      "torch.Size([300, 10])\r\n",
      "300\r\n",
      "  0%|                                                    | 0/62 [00:00<?, ?it/s]\r\n",
      "  0%|                                                    | 0/62 [00:00<?, ?it/s]\u001b[A\r\n",
      "\r\n",
      "  0%|                                                    | 0/38 [00:00<?, ?it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      "100%|██████████████████████████████████████████| 38/38 [00:00<00:00, 242.86it/s]\r\n",
      "  0%|                                                    | 0/62 [00:09<?, ?it/s]\r\n",
      "100%|███████████████████████████████████████████| 38/38 [00:00<00:00, 57.20it/s]\r\n",
      "  0%|                                                    | 0/62 [00:11<?, ?it/s]\r\n",
      "Genomes are linear, segmented, bipartite, and around 21.4kb in total length. The genome is split in two segments RNA1 and RNA2 that together encode three genes. Viral replication is cytoplasmic. Entry into the host cell is achieved by penetration into the host cell. Replication follows the positive stranded RNA virus replication model. Positive stranded RNA virus transcription, using the internal initiation model of subgenomic RNA transcription is the method of transcription. Fish serve as the natural host. Transmission routes are passive diffusion and contact.\r\n",
      "###\r\n",
      "cDNA sequences of viral RNA are synthesized from attenuated master strains by using RT-PCR. This cDNA can then be inserted between an RNA polymerase I (Pol I) promoter and terminator sequence through restriction enzyme digestion. The cDNA and pol I sequence is then, in turn, surrounded by an RNA polymerase II (Pol II) promoter and a polyadenylation site. This entire sequence is then inserted into a plasmid. Six plasmids derived from attenuated master strain cDNA are cotransfected into a target cell, often a chicken egg, alongside two plasmids of the currently circulating wild-type influenza strain. Inside the target cell, the two \"stacked\" Pol I and Pol II enzymes transcribe the viral cDNA to synthesize both negative-sense viral RNA and positive-sense mRNA, effectively creating an attenuated virus. The result is a defective vaccine strain that is similar to the current virus strain, allowing a host to build immunity. This synthesized vaccine strain can then be used as a seed virus to create further vaccines.\r\n",
      "###\r\n",
      "\"Petuviruses\" (including other members of the \"Caulimovirusus\" such as \"soymoviruses\" and \"Cavemoviruses\") form isometric particles whereas members of \"Badnaviruses\" and \"Tungroviruses\" have bacilliform virus particles. Viral replication is nuclear/cytoplasmic. Entry into the host cell is achieved by attachment of the viral proteins to host receptors, which mediates endocytosis. Replication follows the dsDNA(RT) replication model. DNA-templated transcription, specifically dsDNA(RT) transcription is the method of transcription. The virus exits the host cell by nuclear pore export, and tubule-guided viral movement. Plants serve as the natural host. Transmission routes are mechanical and grafting.\r\n",
      "###\r\n",
      "In the field of virology, reverse-genetics techniques can be used to recover full-length infectious viruses with desired mutations or insertions in the viral genomes or in specific virus genes. Technologies that allow these manipulations include circular polymerase extension reaction (CPER) which was first used to generate infectious cDNA for Kunjin virus a close relative of West Nile virus. CPER has also been successfully utilised to generate a range of positive-sense RNA viruses such as SARS-CoV-2, the causative agent of COVID-19.\r\n",
      "###\r\n",
      "As orthomyxoviruses do not encode a capping enzyme, initiation of transcription involves the virus cutting the cap off the 5′-end of host mRNAs, so that the mRNA is recognised by the host translation machinery. A similar \"cap snatching\" process is used by other orthomyxoviruses, but a much longer host RNA sequence is cleaved along with the cap and incorporated into the viral mRNA.\r\n",
      "\r\n",
      "V2:\r\n",
      "\r\n",
      "Context 4: Genomes are linear, segmented, bipartite, and around 21.4kb in total length. The genome is split in two segments RNA1 and RNA2 that together encode three genes. Viral replication is cytoplasmic. Entry into the host cell is achieved by penetration into the host cell. Replication follows the positive stranded RNA virus replication model. Positive stranded RNA virus transcription, using the internal initiation model of subgenomic RNA transcription is the method of transcription. Fish serve as the natural host. Transmission routes are passive diffusion and contact.\r\n",
      "###\r\n",
      "Context 3: cDNA sequences of viral RNA are synthesized from attenuated master strains by using RT-PCR. This cDNA can then be inserted between an RNA polymerase I (Pol I) promoter and terminator sequence through restriction enzyme digestion. The cDNA and pol I sequence is then, in turn, surrounded by an RNA polymerase II (Pol II) promoter and a polyadenylation site. This entire sequence is then inserted into a plasmid. Six plasmids derived from attenuated master strain cDNA are cotransfected into a target cell, often a chicken egg, alongside two plasmids of the currently circulating wild-type influenza strain. Inside the target cell, the two \"stacked\" Pol I and Pol II enzymes transcribe the viral cDNA to synthesize both negative-sense viral RNA and positive-sense mRNA, effectively creating an attenuated virus. The result is a defective vaccine strain that is similar to the current virus strain, allowing a host to build immunity. This synthesized vaccine strain can then be used as a seed virus to create further vaccines.\r\n",
      "###\r\n",
      "Context 2: \"Petuviruses\" (including other members of the \"Caulimovirusus\" such as \"soymoviruses\" and \"Cavemoviruses\") form isometric particles whereas members of \"Badnaviruses\" and \"Tungroviruses\" have bacilliform virus particles. Viral replication is nuclear/cytoplasmic. Entry into the host cell is achieved by attachment of the viral proteins to host receptors, which mediates endocytosis. Replication follows the dsDNA(RT) replication model. DNA-templated transcription, specifically dsDNA(RT) transcription is the method of transcription. The virus exits the host cell by nuclear pore export, and tubule-guided viral movement. Plants serve as the natural host. Transmission routes are mechanical and grafting.\r\n",
      "###\r\n",
      "Context 1: In the field of virology, reverse-genetics techniques can be used to recover full-length infectious viruses with desired mutations or insertions in the viral genomes or in specific virus genes. Technologies that allow these manipulations include circular polymerase extension reaction (CPER) which was first used to generate infectious cDNA for Kunjin virus a close relative of West Nile virus. CPER has also been successfully utilised to generate a range of positive-sense RNA viruses such as SARS-CoV-2, the causative agent of COVID-19.\r\n",
      "###\r\n",
      "Context 0: As orthomyxoviruses do not encode a capping enzyme, initiation of transcription involves the virus cutting the cap off the 5′-end of host mRNAs, so that the mRNA is recognised by the host translation machinery. A similar \"cap snatching\" process is used by other orthomyxoviruses, but a much longer host RNA sequence is cleaved along with the cap and incorporated into the viral mRNA.\r\n",
      "\r\n",
      "DEBERTA:\r\n",
      "\r\n",
      "As orthomyxoviruses do not encode a capping enzyme, initiation of transcription involves the virus cutting the cap off the 5′-end of host mRNAs, so that the mRNA is recognised by the host translation machinery. A similar \"cap snatching\" process is used by other orthomyxoviruses, but a much longer host RNA sequence is cleaved along with the cap and incorporated into the viral mRNA. [SEP] In the field of virology, reverse-genetics techniques can be used to recover full-length infectious viruses with desired mutations or insertions in the viral genomes or in specific virus genes. Technologies that allow these manipulations include circular polymerase extension reaction (CPER) which was first used to generate infectious cDNA for Kunjin virus a close relative of West Nile virus. CPER has also been successfully utilised to generate a range of positive-sense RNA viruses such as SARS-CoV-2, the causative agent of COVID-19. [SEP] \"Petuviruses\" (including other members of the \"Caulimovirusus\" such as \"soymoviruses\" and \"Cavemoviruses\") form isometric particles whereas members of \"Badnaviruses\" and \"Tungroviruses\" have bacilliform virus particles. Viral replication is nuclear/cytoplasmic. Entry into the host cell is achieved by attachment of the viral proteins to host receptors, which mediates endocytosis. Replication follows the dsDNA(RT) replication model. DNA-templated transcription, specifically dsDNA(RT) transcription is the method of transcription. The virus exits the host cell by nuclear pore export, and tubule-guided viral movement. Plants serve as the natural host. Transmission routes are mechanical and grafting. [SEP] cDNA sequences of viral RNA are synthesized from attenuated master strains by using RT-PCR. This cDNA can then be inserted between an RNA polymerase I (Pol I) promoter and terminator sequence through restriction enzyme digestion. The cDNA and pol I sequence is then, in turn, surrounded by an RNA polymerase II (Pol II) promoter and a polyadenylation site. This entire sequence is then inserted into a plasmid. Six plasmids derived from attenuated master strain cDNA are cotransfected into a target cell, often a chicken egg, alongside two plasmids of the currently circulating wild-type influenza strain. Inside the target cell, the two \"stacked\" Pol I and Pol II enzymes transcribe the viral cDNA to synthesize both negative-sense viral RNA and positive-sense mRNA, effectively creating an attenuated virus. The result is a defective vaccine strain that is similar to the current virus strain, allowing a host to build immunity. This synthesized vaccine strain can then be used as a seed virus to create further vaccines. [SEP] Genomes are linear, segmented, bipartite, and around 21.4kb in total length. The genome is split in two segments RNA1 and RNA2 that together encode three genes. Viral replication is cytoplasmic. Entry into the host cell is achieved by penetration into the host cell. Replication follows the positive stranded RNA virus replication model. Positive stranded RNA virus transcription, using the internal initiation model of subgenomic RNA transcription is the method of transcription. Fish serve as the natural host. Transmission routes are passive diffusion and contact.\r\n"
     ]
    }
   ],
   "source": [
    "!python get_topk.py --wiki \"new\" --model_name \"gte-large\" --test_file \"test3\" --topk 5 --ind 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fa212419",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:05:46.678477Z",
     "iopub.status.busy": "2023-10-10T12:05:46.677767Z",
     "iopub.status.idle": "2023-10-10T12:06:21.889085Z",
     "shell.execute_reply": "2023-10-10T12:06:21.887778Z"
    },
    "papermill": {
     "duration": 35.234456,
     "end_time": "2023-10-10T12:06:21.891483",
     "exception": false,
     "start_time": "2023-10-10T12:05:46.657027",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5\r\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\r\n",
      "Using cirrus wiki\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']\r\n",
      "  warnings.warn(f\"unable to load libtensorflow_io_plugins.so: {e}\")\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tensorflow13GcsFileSystemE']\r\n",
      "  warnings.warn(f\"file system plugins are not loaded: {e}\")\r\n",
      "100%|█████████████████████████████████████████| 300/300 [00:05<00:00, 53.20it/s]\r\n",
      "torch.Size([300, 768])\r\n",
      "torch.Size([300, 10])\r\n",
      "300\r\n",
      "  0%|                                                   | 0/117 [00:00<?, ?it/s]\r\n",
      "  0%|                                                   | 0/118 [00:00<?, ?it/s]\u001b[A\r\n",
      "\r\n",
      "  0%|                                                    | 0/38 [00:00<?, ?it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      "100%|██████████████████████████████████████████| 38/38 [00:00<00:00, 286.03it/s]\r\n",
      "  0%|                                                   | 0/117 [00:10<?, ?it/s]\r\n",
      "100%|███████████████████████████████████████████| 38/38 [00:00<00:00, 61.86it/s]\r\n",
      "  0%|                                                   | 0/118 [00:13<?, ?it/s]\r\n",
      "Viral replication is cytoplasmic. Entry into the host cell is achieved by attachment of the viral envelope protein E to host receptors, which mediates clathrin-mediated endocytosis. Replication follows the positive-stranded RNA virus replication model. Positive-stranded RNA virus transcription is the method of transcription. Translation takes place by viral initiation. The virus exits the host cell by budding. Humans and mammals serve as the natural hosts. The virus is transmitted via vectors (ticks and mosquitoes).\r\n",
      "###\r\n",
      "As in all orthomyxoviruses, the largest three segments (1–3) encode the three subunits of the RNA polymerase. In thogotoviruses, segment 4 encodes the glycoprotein and segment 5 the nucleoprotein. The messenger RNA (mRNA) from segment 6 can be spliced to encode the matrix protein or unspliced to encode ML, which has 38 additional amino acids at its C-terminus.\r\n",
      "###\r\n",
      "The procedure used was devised by Garret Ihler and Karin Ippen. Complementary strands of the gene, carried within non-complementary strands of viral DNA, from viruses transducing the gene in reverse orientations which had been separated by a poly U,G density-gradient technique, were annealed to form double-stranded DNA.\r\n",
      "\r\n",
      "V2:\r\n",
      "\r\n",
      "Context 4: Specifically, this method covalently links proteins of interest to the mRNAs encoding them, then detects the mRNA pieces using reverse transcription PCRs. The mRNA may then be amplified and sequenced. The combined method was titled IVV-HiTSeq and can be performed under cell-free conditions, though its results may not be representative of in vivo conditions.\r\n",
      "###\r\n",
      "Context 3: Following binding of viral envelope glycoproteins to cell membrane receptors, the virion is internalized and dismantled, allowing viral DNA to migrate to the cell nucleus. Within the nucleus, replication of viral DNA and transcription of viral genes occurs.[citation needed] During symptomatic infection, infected cells transcribe lytic viral genes.\r\n",
      "###\r\n",
      "Context 2: Viral replication is cytoplasmic. Entry into the host cell is achieved by attachment of the viral envelope protein E to host receptors, which mediates clathrin-mediated endocytosis. Replication follows the positive-stranded RNA virus replication model. Positive-stranded RNA virus transcription is the method of transcription. Translation takes place by viral initiation. The virus exits the host cell by budding. Humans and mammals serve as the natural hosts. The virus is transmitted via vectors (ticks and mosquitoes).\r\n",
      "###\r\n",
      "Context 1: As in all orthomyxoviruses, the largest three segments (1–3) encode the three subunits of the RNA polymerase. In thogotoviruses, segment 4 encodes the glycoprotein and segment 5 the nucleoprotein. The messenger RNA (mRNA) from segment 6 can be spliced to encode the matrix protein or unspliced to encode ML, which has 38 additional amino acids at its C-terminus.\r\n",
      "###\r\n",
      "Context 0: The procedure used was devised by Garret Ihler and Karin Ippen. Complementary strands of the gene, carried within non-complementary strands of viral DNA, from viruses transducing the gene in reverse orientations which had been separated by a poly U,G density-gradient technique, were annealed to form double-stranded DNA.\r\n",
      "\r\n",
      "DEBERTA:\r\n",
      "\r\n",
      "The procedure used was devised by Garret Ihler and Karin Ippen. Complementary strands of the gene, carried within non-complementary strands of viral DNA, from viruses transducing the gene in reverse orientations which had been separated by a poly U,G density-gradient technique, were annealed to form double-stranded DNA. [SEP] As in all orthomyxoviruses, the largest three segments (1–3) encode the three subunits of the RNA polymerase. In thogotoviruses, segment 4 encodes the glycoprotein and segment 5 the nucleoprotein. The messenger RNA (mRNA) from segment 6 can be spliced to encode the matrix protein or unspliced to encode ML, which has 38 additional amino acids at its C-terminus. [SEP] Viral replication is cytoplasmic. Entry into the host cell is achieved by attachment of the viral envelope protein E to host receptors, which mediates clathrin-mediated endocytosis. Replication follows the positive-stranded RNA virus replication model. Positive-stranded RNA virus transcription is the method of transcription. Translation takes place by viral initiation. The virus exits the host cell by budding. Humans and mammals serve as the natural hosts. The virus is transmitted via vectors (ticks and mosquitoes).\r\n"
     ]
    }
   ],
   "source": [
    "!python get_topk.py --wiki \"cirrus\" --model_name \"gte-base\" --test_file \"test4\" --topk 3 --ind 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5468b891",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:06:21.939598Z",
     "iopub.status.busy": "2023-10-10T12:06:21.939006Z",
     "iopub.status.idle": "2023-10-10T12:07:29.859218Z",
     "shell.execute_reply": "2023-10-10T12:07:29.857989Z"
    },
    "papermill": {
     "duration": 67.947012,
     "end_time": "2023-10-10T12:07:29.861499",
     "exception": false,
     "start_time": "2023-10-10T12:06:21.914487",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5\r\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\r\n",
      "Using cirrus wiki\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']\r\n",
      "  warnings.warn(f\"unable to load libtensorflow_io_plugins.so: {e}\")\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tensorflow13GcsFileSystemE']\r\n",
      "  warnings.warn(f\"file system plugins are not loaded: {e}\")\r\n",
      "100%|█████████████████████████████████████████| 300/300 [00:12<00:00, 23.25it/s]\r\n",
      "torch.Size([300, 1024])\r\n",
      "torch.Size([300, 10])\r\n",
      "300\r\n",
      "  0%|                                                   | 0/117 [00:00<?, ?it/s]\r\n",
      "  0%|                                                   | 0/118 [00:00<?, ?it/s]\u001b[A\r\n",
      "\r\n",
      "  0%|                                                    | 0/38 [00:00<?, ?it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      "100%|██████████████████████████████████████████| 38/38 [00:00<00:00, 269.65it/s]\r\n",
      "  0%|                                                   | 0/117 [00:11<?, ?it/s]\r\n",
      "100%|███████████████████████████████████████████| 38/38 [00:00<00:00, 60.14it/s]\r\n",
      "  0%|                                                   | 0/118 [00:14<?, ?it/s]\r\n",
      "See Herpesvirales#Taxonomy for information on taxonomic history, phylogenetic research, and the nomenclatural system. All members of the Herpesviridae share a common structure; a relatively large, monopartite, double-stranded, linear DNA genome encoding 100-200 genes encased within an icosahedral protein cage (with T=16 symmetry) called the capsid, which is itself wrapped in a protein layer called the tegument containing both viral proteins and viral mRNAs and a lipid bilayer membrane called the envelope.\r\n",
      "###\r\n",
      "Thogotovirus is a genus of enveloped RNA viruses, one of seven genera in the virus family Orthomyxoviridae. Their single-stranded, negative-sense RNA genome has six or seven segments. Thogotoviruses are distinguished from most other orthomyxoviruses by being arboviruses – viruses that are transmitted by arthropods, in this case usually ticks. Thogotoviruses can replicate in both tick cells and vertebrate cells; one subtype has also been isolated from mosquitoes.\r\n",
      "###\r\n",
      "Some notable pathogenic viruses cause smallpox, influenza, mumps, measles, chickenpox, ebola, and rubella. Viruses typically range between 20 and 300 nanometers in length. This type of pathogen is not cellular, and is instead composed of either RNA (Ribonucleic acid) or DNA (Deoxyribonucleic acid) within a protein shell - the capsid. Pathogenic viruses infiltrate host cells and manipulate the organelles within the cell such as the Ribosomes, Golgi Apparatus, and Endoplasmic Reticulum in order to multiply which commonly results in the death of the host cell via cellular decay.\r\n",
      "###\r\n",
      "The icosahedral nucleocapsid is similar to that of tailed bacteriophage in the order Caudovirales. This capsid has 161 capsomers consisting of 150 hexons and 11 pentons, as well as a portal complex that allows entry and exit of DNA into the capsid. All herpesviruses are nuclear-replicating—the viral DNA is transcribed to mRNA within the infected cell's nucleus.[citation needed] Infection is initiated when a viral particle contacts a cell with specific types of receptor molecules on the cell surface.\r\n",
      "###\r\n",
      "Viral replication is cytoplasmic. Entry into the host cell is achieved by attachment of the viral envelope protein E to host receptors, which mediates clathrin-mediated endocytosis. Replication follows the positive-stranded RNA virus replication model. Positive-stranded RNA virus transcription is the method of transcription. Translation takes place by viral initiation. The virus exits the host cell by budding. Humans and mammals serve as the natural hosts. The virus is transmitted via vectors (ticks and mosquitoes).\r\n",
      "\r\n",
      "V2:\r\n",
      "\r\n",
      "Context 4: See Herpesvirales#Taxonomy for information on taxonomic history, phylogenetic research, and the nomenclatural system. All members of the Herpesviridae share a common structure; a relatively large, monopartite, double-stranded, linear DNA genome encoding 100-200 genes encased within an icosahedral protein cage (with T=16 symmetry) called the capsid, which is itself wrapped in a protein layer called the tegument containing both viral proteins and viral mRNAs and a lipid bilayer membrane called the envelope.\r\n",
      "###\r\n",
      "Context 3: Thogotovirus is a genus of enveloped RNA viruses, one of seven genera in the virus family Orthomyxoviridae. Their single-stranded, negative-sense RNA genome has six or seven segments. Thogotoviruses are distinguished from most other orthomyxoviruses by being arboviruses – viruses that are transmitted by arthropods, in this case usually ticks. Thogotoviruses can replicate in both tick cells and vertebrate cells; one subtype has also been isolated from mosquitoes.\r\n",
      "###\r\n",
      "Context 2: Some notable pathogenic viruses cause smallpox, influenza, mumps, measles, chickenpox, ebola, and rubella. Viruses typically range between 20 and 300 nanometers in length. This type of pathogen is not cellular, and is instead composed of either RNA (Ribonucleic acid) or DNA (Deoxyribonucleic acid) within a protein shell - the capsid. Pathogenic viruses infiltrate host cells and manipulate the organelles within the cell such as the Ribosomes, Golgi Apparatus, and Endoplasmic Reticulum in order to multiply which commonly results in the death of the host cell via cellular decay.\r\n",
      "###\r\n",
      "Context 1: The icosahedral nucleocapsid is similar to that of tailed bacteriophage in the order Caudovirales. This capsid has 161 capsomers consisting of 150 hexons and 11 pentons, as well as a portal complex that allows entry and exit of DNA into the capsid. All herpesviruses are nuclear-replicating—the viral DNA is transcribed to mRNA within the infected cell's nucleus.[citation needed] Infection is initiated when a viral particle contacts a cell with specific types of receptor molecules on the cell surface.\r\n",
      "###\r\n",
      "Context 0: Viral replication is cytoplasmic. Entry into the host cell is achieved by attachment of the viral envelope protein E to host receptors, which mediates clathrin-mediated endocytosis. Replication follows the positive-stranded RNA virus replication model. Positive-stranded RNA virus transcription is the method of transcription. Translation takes place by viral initiation. The virus exits the host cell by budding. Humans and mammals serve as the natural hosts. The virus is transmitted via vectors (ticks and mosquitoes).\r\n",
      "\r\n",
      "DEBERTA:\r\n",
      "\r\n",
      "Viral replication is cytoplasmic. Entry into the host cell is achieved by attachment of the viral envelope protein E to host receptors, which mediates clathrin-mediated endocytosis. Replication follows the positive-stranded RNA virus replication model. Positive-stranded RNA virus transcription is the method of transcription. Translation takes place by viral initiation. The virus exits the host cell by budding. Humans and mammals serve as the natural hosts. The virus is transmitted via vectors (ticks and mosquitoes). [SEP] The icosahedral nucleocapsid is similar to that of tailed bacteriophage in the order Caudovirales. This capsid has 161 capsomers consisting of 150 hexons and 11 pentons, as well as a portal complex that allows entry and exit of DNA into the capsid. All herpesviruses are nuclear-replicating—the viral DNA is transcribed to mRNA within the infected cell's nucleus.[citation needed] Infection is initiated when a viral particle contacts a cell with specific types of receptor molecules on the cell surface. [SEP] Some notable pathogenic viruses cause smallpox, influenza, mumps, measles, chickenpox, ebola, and rubella. Viruses typically range between 20 and 300 nanometers in length. This type of pathogen is not cellular, and is instead composed of either RNA (Ribonucleic acid) or DNA (Deoxyribonucleic acid) within a protein shell - the capsid. Pathogenic viruses infiltrate host cells and manipulate the organelles within the cell such as the Ribosomes, Golgi Apparatus, and Endoplasmic Reticulum in order to multiply which commonly results in the death of the host cell via cellular decay. [SEP] Thogotovirus is a genus of enveloped RNA viruses, one of seven genera in the virus family Orthomyxoviridae. Their single-stranded, negative-sense RNA genome has six or seven segments. Thogotoviruses are distinguished from most other orthomyxoviruses by being arboviruses – viruses that are transmitted by arthropods, in this case usually ticks. Thogotoviruses can replicate in both tick cells and vertebrate cells; one subtype has also been isolated from mosquitoes. [SEP] See Herpesvirales#Taxonomy for information on taxonomic history, phylogenetic research, and the nomenclatural system. All members of the Herpesviridae share a common structure; a relatively large, monopartite, double-stranded, linear DNA genome encoding 100-200 genes encased within an icosahedral protein cage (with T=16 symmetry) called the capsid, which is itself wrapped in a protein layer called the tegument containing both viral proteins and viral mRNAs and a lipid bilayer membrane called the envelope.\r\n"
     ]
    }
   ],
   "source": [
    "!python get_topk.py --wiki \"cirrus\" --model_name \"e5-large\" --test_file \"test5\" --topk 5 --ind 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7014c407",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:07:29.917250Z",
     "iopub.status.busy": "2023-10-10T12:07:29.916479Z",
     "iopub.status.idle": "2023-10-10T12:09:13.115126Z",
     "shell.execute_reply": "2023-10-10T12:09:13.114026Z"
    },
    "papermill": {
     "duration": 103.228902,
     "end_time": "2023-10-10T12:09:13.117374",
     "exception": false,
     "start_time": "2023-10-10T12:07:29.888472",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5\r\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\r\n",
      "Using cirrus wiki\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']\r\n",
      "  warnings.warn(f\"unable to load libtensorflow_io_plugins.so: {e}\")\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tensorflow13GcsFileSystemE']\r\n",
      "  warnings.warn(f\"file system plugins are not loaded: {e}\")\r\n",
      "100%|█████████████████████████████████████████| 300/300 [00:20<00:00, 14.83it/s]\r\n",
      "torch.Size([1500, 1024])\r\n",
      "torch.Size([300, 25])\r\n",
      "300\r\n",
      "  0%|                                                   | 0/117 [00:00<?, ?it/s]\r\n",
      "  0%|                                                   | 0/118 [00:00<?, ?it/s]\u001b[A\r\n",
      "\r\n",
      "  0%|                                                   | 0/188 [00:00<?, ?it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      " 12%|█████                                    | 23/188 [00:00<00:00, 228.85it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      " 28%|███████████▎                             | 52/188 [00:00<00:00, 260.72it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      " 45%|██████████████████▌                      | 85/188 [00:00<00:00, 291.02it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      " 61%|████████████████████████▍               | 115/188 [00:00<00:00, 268.98it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      " 79%|███████████████████████████████▍        | 148/188 [00:00<00:00, 289.36it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      "100%|████████████████████████████████████████| 188/188 [00:00<00:00, 289.28it/s]\r\n",
      "\r\n",
      "\r\n",
      "  0%|                                                   | 0/188 [00:00<?, ?it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      "  1%|▏                                          | 1/188 [00:00<02:06,  1.48it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      " 15%|██████▎                                   | 28/188 [00:00<00:03, 48.13it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      " 30%|████████████▋                             | 57/188 [00:00<00:01, 95.89it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      " 46%|██████████████████▉                      | 87/188 [00:00<00:00, 140.60it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      " 62%|████████████████████████▉               | 117/188 [00:01<00:00, 178.43it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      " 79%|███████████████████████████████▋        | 149/188 [00:01<00:00, 213.50it/s]\u001b[A\u001b[A\r\n",
      "\r\n",
      "100%|████████████████████████████████████████| 188/188 [00:01<00:00, 143.69it/s]\r\n",
      "  0%|                                                   | 0/117 [00:42<?, ?it/s]\r\n",
      "  0%|                                                   | 0/118 [00:44<?, ?it/s]\r\n",
      "Viral replication is cytoplasmic. Entry into the host cell is achieved by attachment of the viral envelope protein E to host receptors, which mediates clathrin-mediated endocytosis. Replication follows the positive-stranded RNA virus replication model. Positive-stranded RNA virus transcription is the method of transcription. Translation takes place by viral initiation. The virus exits the host cell by budding. Humans and mammals serve as the natural hosts. The virus is transmitted via vectors (ticks and mosquitoes).\r\n",
      "###\r\n",
      "This method is based on use of RNA polymerase (RNAP), which is attached to a polystyrene bead. One end of DNA to be sequenced is attached to another bead, with both beads being placed in optical traps. RNAP motion during transcription brings the beads in closer and their relative distance changes, which can then be recorded at a single nucleotide resolution.\r\n",
      "###\r\n",
      "Proc. Natl. Acad. Sci. U.S.A. 70 (12): 3581–84. Bibcode:1973PNAS...70.3581G. doi:10.1073/pnas.70.12.3581. PMC 427284. PMID 4587255. Sanger F, Air GM, Barrell BG, Brown NL, Coulson AR, Fiddes CA, Hutchison CA, Slocombe PM, Smith M (February 1977). \"Nucleotide sequence of bacteriophage phi X174 DNA\". Nature. 265 (5596): 687–95. Bibcode:1977Natur.265..687S. doi:10.1038/265687a0. PMID 870828. S2CID 4206886. Marks, L. \"The next frontier: Human viruses\". What is Biotechnology?. Retrieved 27 June 2023.\r\n",
      "###\r\n",
      "(2008). \"Accurate whole human genome sequencing using reversible terminator chemistry\". Nature. 456 (7218): 53–59. Bibcode:2008Natur.456...53B. doi:10.1038/nature07517. PMC 2581791. PMID 18987734. Canard B, Sarfati S (13 October 1994), Novel derivatives usable for the sequencing of nucleic acids, retrieved 9 March 2016 Canard B, Sarfati RS (October 1994).\r\n",
      "###\r\n",
      "The success of any DNA sequencing protocol relies upon the DNA or RNA sample extraction and preparation from the biological material of interest. A successful DNA extraction will yield a DNA sample with long, non-degraded strands. A successful RNA extraction will yield a RNA sample that should be converted to complementary DNA (cDNA) using reverse transcriptase—a DNA polymerase that synthesizes a complementary DNA based on existing strands of RNA in a PCR-like manner.\r\n",
      "\r\n",
      "V2:\r\n",
      "\r\n",
      "Context 4: Viral replication is cytoplasmic. Entry into the host cell is achieved by attachment of the viral envelope protein E to host receptors, which mediates clathrin-mediated endocytosis. Replication follows the positive-stranded RNA virus replication model. Positive-stranded RNA virus transcription is the method of transcription. Translation takes place by viral initiation. The virus exits the host cell by budding. Humans and mammals serve as the natural hosts. The virus is transmitted via vectors (ticks and mosquitoes).\r\n",
      "###\r\n",
      "Context 3: This method is based on use of RNA polymerase (RNAP), which is attached to a polystyrene bead. One end of DNA to be sequenced is attached to another bead, with both beads being placed in optical traps. RNAP motion during transcription brings the beads in closer and their relative distance changes, which can then be recorded at a single nucleotide resolution.\r\n",
      "###\r\n",
      "Context 2: Proc. Natl. Acad. Sci. U.S.A. 70 (12): 3581–84. Bibcode:1973PNAS...70.3581G. doi:10.1073/pnas.70.12.3581. PMC 427284. PMID 4587255. Sanger F, Air GM, Barrell BG, Brown NL, Coulson AR, Fiddes CA, Hutchison CA, Slocombe PM, Smith M (February 1977). \"Nucleotide sequence of bacteriophage phi X174 DNA\". Nature. 265 (5596): 687–95. Bibcode:1977Natur.265..687S. doi:10.1038/265687a0. PMID 870828. S2CID 4206886. Marks, L. \"The next frontier: Human viruses\". What is Biotechnology?. Retrieved 27 June 2023.\r\n",
      "###\r\n",
      "Context 1: (2008). \"Accurate whole human genome sequencing using reversible terminator chemistry\". Nature. 456 (7218): 53–59. Bibcode:2008Natur.456...53B. doi:10.1038/nature07517. PMC 2581791. PMID 18987734. Canard B, Sarfati S (13 October 1994), Novel derivatives usable for the sequencing of nucleic acids, retrieved 9 March 2016 Canard B, Sarfati RS (October 1994).\r\n",
      "###\r\n",
      "Context 0: The success of any DNA sequencing protocol relies upon the DNA or RNA sample extraction and preparation from the biological material of interest. A successful DNA extraction will yield a DNA sample with long, non-degraded strands. A successful RNA extraction will yield a RNA sample that should be converted to complementary DNA (cDNA) using reverse transcriptase—a DNA polymerase that synthesizes a complementary DNA based on existing strands of RNA in a PCR-like manner.\r\n",
      "\r\n",
      "DEBERTA:\r\n",
      "\r\n",
      "The success of any DNA sequencing protocol relies upon the DNA or RNA sample extraction and preparation from the biological material of interest. A successful DNA extraction will yield a DNA sample with long, non-degraded strands. A successful RNA extraction will yield a RNA sample that should be converted to complementary DNA (cDNA) using reverse transcriptase—a DNA polymerase that synthesizes a complementary DNA based on existing strands of RNA in a PCR-like manner. [SEP] (2008). \"Accurate whole human genome sequencing using reversible terminator chemistry\". Nature. 456 (7218): 53–59. Bibcode:2008Natur.456...53B. doi:10.1038/nature07517. PMC 2581791. PMID 18987734. Canard B, Sarfati S (13 October 1994), Novel derivatives usable for the sequencing of nucleic acids, retrieved 9 March 2016 Canard B, Sarfati RS (October 1994). [SEP] Proc. Natl. Acad. Sci. U.S.A. 70 (12): 3581–84. Bibcode:1973PNAS...70.3581G. doi:10.1073/pnas.70.12.3581. PMC 427284. PMID 4587255. Sanger F, Air GM, Barrell BG, Brown NL, Coulson AR, Fiddes CA, Hutchison CA, Slocombe PM, Smith M (February 1977). \"Nucleotide sequence of bacteriophage phi X174 DNA\". Nature. 265 (5596): 687–95. Bibcode:1977Natur.265..687S. doi:10.1038/265687a0. PMID 870828. S2CID 4206886. Marks, L. \"The next frontier: Human viruses\". What is Biotechnology?. Retrieved 27 June 2023. [SEP] This method is based on use of RNA polymerase (RNAP), which is attached to a polystyrene bead. One end of DNA to be sequenced is attached to another bead, with both beads being placed in optical traps. RNAP motion during transcription brings the beads in closer and their relative distance changes, which can then be recorded at a single nucleotide resolution. [SEP] Viral replication is cytoplasmic. Entry into the host cell is achieved by attachment of the viral envelope protein E to host receptors, which mediates clathrin-mediated endocytosis. Replication follows the positive-stranded RNA virus replication model. Positive-stranded RNA virus transcription is the method of transcription. Translation takes place by viral initiation. The virus exits the host cell by budding. Humans and mammals serve as the natural hosts. The virus is transmitted via vectors (ticks and mosquitoes).\r\n"
     ]
    }
   ],
   "source": [
    "!python get_topk.py --wiki \"cirrus\" --model_name \"bge-large\" --test_file \"test6\" --topk 5 --ind 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "157a083e",
   "metadata": {
    "papermill": {
     "duration": 0.032474,
     "end_time": "2023-10-10T12:09:13.183854",
     "exception": false,
     "start_time": "2023-10-10T12:09:13.151380",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3b106893",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:09:13.250021Z",
     "iopub.status.busy": "2023-10-10T12:09:13.249678Z",
     "iopub.status.idle": "2023-10-10T12:09:13.464147Z",
     "shell.execute_reply": "2023-10-10T12:09:13.463189Z"
    },
    "papermill": {
     "duration": 0.249952,
     "end_time": "2023-10-10T12:09:13.465861",
     "exception": false,
     "start_time": "2023-10-10T12:09:13.215909",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import gc\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6413862e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:09:13.533219Z",
     "iopub.status.busy": "2023-10-10T12:09:13.532459Z",
     "iopub.status.idle": "2023-10-10T12:10:42.561423Z",
     "shell.execute_reply": "2023-10-10T12:10:42.560000Z"
    },
    "papermill": {
     "duration": 89.064885,
     "end_time": "2023-10-10T12:10:42.563691",
     "exception": false,
     "start_time": "2023-10-10T12:09:13.498806",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing /kaggle/input/sci-llm-pip-v2/bitsandbytes-0.41.0-py3-none-any.whl\r\n",
      "Installing collected packages: bitsandbytes\r\n",
      "Successfully installed bitsandbytes-0.41.0\r\n",
      "Looking in links: /kaggle/input/transformers-main\r\n",
      "Processing /kaggle/input/transformers-main/transformers-4.34.0-py3-none-any.whl\r\n",
      "Requirement already satisfied: filelock in /opt/conda/lib/python3.10/site-packages (from transformers==4.34.0) (3.12.2)\r\n",
      "Requirement already satisfied: huggingface-hub<1.0,>=0.16.4 in /opt/conda/lib/python3.10/site-packages (from transformers==4.34.0) (0.16.4)\r\n",
      "Requirement already satisfied: numpy>=1.17 in /opt/conda/lib/python3.10/site-packages (from transformers==4.34.0) (1.23.5)\r\n",
      "Requirement already satisfied: packaging>=20.0 in /opt/conda/lib/python3.10/site-packages (from transformers==4.34.0) (21.3)\r\n",
      "Requirement already satisfied: pyyaml>=5.1 in /opt/conda/lib/python3.10/site-packages (from transformers==4.34.0) (6.0)\r\n",
      "Requirement already satisfied: regex!=2019.12.17 in /opt/conda/lib/python3.10/site-packages (from transformers==4.34.0) (2023.6.3)\r\n",
      "Requirement already satisfied: requests in /opt/conda/lib/python3.10/site-packages (from transformers==4.34.0) (2.31.0)\r\n",
      "Processing /kaggle/input/transformers-main/tokenizers-0.14.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (from transformers==4.34.0)\r\n",
      "Requirement already satisfied: safetensors>=0.3.1 in /opt/conda/lib/python3.10/site-packages (from transformers==4.34.0) (0.3.1)\r\n",
      "Requirement already satisfied: tqdm>=4.27 in /opt/conda/lib/python3.10/site-packages (from transformers==4.34.0) (4.65.0)\r\n",
      "Requirement already satisfied: fsspec in /opt/conda/lib/python3.10/site-packages (from huggingface-hub<1.0,>=0.16.4->transformers==4.34.0) (2023.6.0)\r\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in /opt/conda/lib/python3.10/site-packages (from huggingface-hub<1.0,>=0.16.4->transformers==4.34.0) (4.6.3)\r\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /opt/conda/lib/python3.10/site-packages (from packaging>=20.0->transformers==4.34.0) (3.0.9)\r\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /opt/conda/lib/python3.10/site-packages (from requests->transformers==4.34.0) (3.1.0)\r\n",
      "Requirement already satisfied: idna<4,>=2.5 in /opt/conda/lib/python3.10/site-packages (from requests->transformers==4.34.0) (3.4)\r\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /opt/conda/lib/python3.10/site-packages (from requests->transformers==4.34.0) (1.26.15)\r\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.10/site-packages (from requests->transformers==4.34.0) (2023.5.7)\r\n",
      "Installing collected packages: tokenizers, transformers\r\n",
      "  Attempting uninstall: tokenizers\r\n",
      "    Found existing installation: tokenizers 0.13.3\r\n",
      "    Uninstalling tokenizers-0.13.3:\r\n",
      "      Successfully uninstalled tokenizers-0.13.3\r\n",
      "  Attempting uninstall: transformers\r\n",
      "    Found existing installation: transformers 4.30.2\r\n",
      "    Uninstalling transformers-4.30.2:\r\n",
      "      Successfully uninstalled transformers-4.30.2\r\n",
      "Successfully installed tokenizers-0.14.1 transformers-4.34.0\r\n",
      "Processing /kaggle/input/sci-llm-pip-v2/tiktoken-0.4.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\r\n",
      "Requirement already satisfied: regex>=2022.1.18 in /opt/conda/lib/python3.10/site-packages (from tiktoken==0.4.0) (2023.6.3)\r\n",
      "Requirement already satisfied: requests>=2.26.0 in /opt/conda/lib/python3.10/site-packages (from tiktoken==0.4.0) (2.31.0)\r\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /opt/conda/lib/python3.10/site-packages (from requests>=2.26.0->tiktoken==0.4.0) (3.1.0)\r\n",
      "Requirement already satisfied: idna<4,>=2.5 in /opt/conda/lib/python3.10/site-packages (from requests>=2.26.0->tiktoken==0.4.0) (3.4)\r\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /opt/conda/lib/python3.10/site-packages (from requests>=2.26.0->tiktoken==0.4.0) (1.26.15)\r\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.10/site-packages (from requests>=2.26.0->tiktoken==0.4.0) (2023.5.7)\r\n",
      "Installing collected packages: tiktoken\r\n",
      "Successfully installed tiktoken-0.4.0\r\n"
     ]
    }
   ],
   "source": [
    "!pip install /kaggle/input/sci-llm-pip-v2/bitsandbytes-0.41.0-py3-none-any.whl\n",
    "!pip install --no-index --find-links=\"/kaggle/input/transformers-main\" /kaggle/input/transformers-main/transformers-4.34.0-py3-none-any.whl\n",
    "!pip install /kaggle/input/sci-llm-pip-v2/tiktoken-0.4.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "62d80636",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:10:42.689973Z",
     "iopub.status.busy": "2023-10-10T12:10:42.689572Z",
     "iopub.status.idle": "2023-10-10T12:10:42.694510Z",
     "shell.execute_reply": "2023-10-10T12:10:42.693516Z"
    },
    "papermill": {
     "duration": 0.046776,
     "end_time": "2023-10-10T12:10:42.696317",
     "exception": false,
     "start_time": "2023-10-10T12:10:42.649541",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "os.environ[\"TIKTOKEN_CACHE_DIR\"] = \"/kaggle/input/tiktoken-cache\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "14ba4fbe",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:10:42.769209Z",
     "iopub.status.busy": "2023-10-10T12:10:42.768721Z",
     "iopub.status.idle": "2023-10-10T12:10:42.779459Z",
     "shell.execute_reply": "2023-10-10T12:10:42.778287Z"
    },
    "papermill": {
     "duration": 0.049565,
     "end_time": "2023-10-10T12:10:42.781266",
     "exception": false,
     "start_time": "2023-10-10T12:10:42.731701",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing inference.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile inference.py\n",
    "\n",
    "\n",
    "from transformers import AutoModelForCausalLM, AutoTokenizer\n",
    "import itertools\n",
    "import random\n",
    "import argparse\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from string import Template\n",
    "from pathlib import Path\n",
    "import time\n",
    "import torch\n",
    "from tqdm.auto import tqdm\n",
    "import gc\n",
    "import itertools\n",
    "\n",
    "#os.environ[\"TOKENIZERS_PARALLELISM\"] = \"false\"\n",
    "\n",
    "def precision_at_k(r, k):\n",
    "    \"\"\"Precision at k\"\"\"\n",
    "    assert k <= len(r)\n",
    "    assert k != 0\n",
    "    return sum(int(x) for x in r[:k]) / k\n",
    "\n",
    "def MAP_at_3(predictions, true_items):\n",
    "    \"\"\"Score is mean average precision at 3\"\"\"\n",
    "    U = len(predictions)\n",
    "    map_at_3 = 0.0\n",
    "    for u in range(U):\n",
    "        user_preds = predictions[u]\n",
    "        user_true = true_items[u]\n",
    "        user_results = [1 if item == user_true else 0 for item in user_preds]\n",
    "        for k in range(min(len(user_preds), 3)):\n",
    "            map_at_3 += precision_at_k(user_results, k+1) * user_results[k]\n",
    "    return map_at_3 / U\n",
    "\n",
    "def longest_common_prefix(strs):\n",
    "    if not strs:\n",
    "        return \"\"\n",
    "\n",
    "    shortest = min(strs,key=len)\n",
    "\n",
    "    for i, char in enumerate(shortest):\n",
    "        for other in strs:\n",
    "            if other[i] != char:\n",
    "                return shortest[:i]\n",
    "\n",
    "    return shortest \n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    \n",
    "    ap = argparse.ArgumentParser()\n",
    "    ap.add_argument('--device', type=str, required=True)\n",
    "    ap.add_argument('--model_name', type=str, required=True)\n",
    "    ap.add_argument('--quantization', type=int, required=True)\n",
    "    ap.add_argument('--model_type', type=int, required=True)\n",
    "    ap.add_argument('--test_file', type=str, required=True)\n",
    "    args = ap.parse_args()\n",
    "    \n",
    "    if args.device == \"auto\":\n",
    "        DEVICE_MAP = \"auto\"\n",
    "        DEVICE = \"cuda:0\"\n",
    "    else:\n",
    "        DEVICE_MAP = {\"\":args.device}\n",
    "        DEVICE = args.device\n",
    "    \n",
    "    llm_backbone = args.model_name\n",
    "    \n",
    "    test = pd.read_parquet(args.test_file).reset_index(drop=True)\n",
    "\n",
    "\n",
    "    new_obs = []\n",
    "    for idx, row in test.iterrows():\n",
    "        for opt in \"ABCDE\":\n",
    "            new_obs.append((row[\"id\"], row[\"context\"], row[\"context_v2\"], row[\"prompt\"], row[opt]))\n",
    "    df = pd.DataFrame(new_obs, columns=[\"id\", \"context\", \"context_v2\", \"question\", \"answer\"])\n",
    "\n",
    "    \n",
    "    print(df.head(1))\n",
    "    \n",
    "\n",
    "    tokenizer = AutoTokenizer.from_pretrained(\n",
    "        llm_backbone,\n",
    "        use_fast=True,\n",
    "        trust_remote_code=True,\n",
    "        padding_side=\"right\",\n",
    "        truncation_side=\"left\"\n",
    "    )\n",
    "    \n",
    "    if tokenizer.pad_token is None:\n",
    "        if tokenizer.unk_token is not None:\n",
    "            tokenizer.pad_token = tokenizer.unk_token\n",
    "        else:\n",
    "            tokenizer.pad_token = tokenizer.eos_token\n",
    "    \n",
    "    from transformers import BitsAndBytesConfig\n",
    "    \n",
    "    \n",
    "#     quantization_config = BitsAndBytesConfig(\n",
    "#        load_in_4bit=True,\n",
    "#        bnb_4bit_quant_type=\"nf4\",\n",
    "#        bnb_4bit_use_double_quant=False,\n",
    "#        bnb_4bit_compute_dtype=torch.float16\n",
    "#     )\n",
    "\n",
    "    if args.quantization == 0:\n",
    "        quantization_config = None\n",
    "\n",
    "    elif args.quantization == 1:\n",
    "        quantization_config = BitsAndBytesConfig(\n",
    "            load_in_8bit=True,\n",
    "            llm_int8_threshold=0.0,\n",
    "        )\n",
    "\n",
    "\n",
    "    model = AutoModelForCausalLM.from_pretrained(\n",
    "        llm_backbone,\n",
    "        torch_dtype=torch.float16,\n",
    "        quantization_config=quantization_config,\n",
    "        device_map=DEVICE_MAP,\n",
    "        low_cpu_mem_usage=True,\n",
    "        trust_remote_code=True\n",
    "    ).eval()\n",
    "\n",
    "    head_weights = torch.load(llm_backbone + \"/head.pth\", map_location=\"cpu\")\n",
    "    hidden_size = head_weights.shape[1]\n",
    "    print(hidden_size)\n",
    "    # hidden_size = 32000\n",
    "    head = torch.nn.Linear(hidden_size, 1, bias=False)\n",
    "\n",
    "    head.weight.data = head_weights\n",
    "\n",
    "    head.to(DEVICE).eval()\n",
    "\n",
    "    model.config.pad_token_id = tokenizer.pad_token_id\n",
    "    \n",
    "    gc.collect()\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "    import itertools\n",
    "    import random\n",
    "    maps = []\n",
    "\n",
    "    progress_bar = tqdm(df.iterrows(), total=len(df))\n",
    "\n",
    "    preds = []\n",
    "    instructions = []\n",
    "    pooled = []\n",
    "    past_key_values = None\n",
    "    for idx,row in progress_bar:\n",
    "        \n",
    "        inst = f\"Answer: {row['answer']}\\n###\\nIs this answer correct? \"\n",
    "        #inst = \"a b c d\"*10_000\n",
    "        instructions.append(inst)\n",
    "        \n",
    "        if idx % 5 == 0:\n",
    "            \n",
    "            if past_key_values is not None:\n",
    "                del past_key_values\n",
    "            \n",
    "            preprompt = f\"{row['context_v2']}\\n###\\nQuestion: {row['question']}\\n###\\n\"\n",
    "            #preprompt = \"a b c d\"*10_000\n",
    "            inputs = tokenizer(preprompt, return_tensors='pt', add_special_tokens=False, truncation=True, padding=\"longest\", max_length=1024)\n",
    "            \n",
    "            tok_length = inputs[\"input_ids\"].shape[1] + tokenizer(instructions, return_tensors='pt', add_special_tokens=False, truncation=True, padding=\"longest\", max_length=1024)[\"input_ids\"].shape[1]\n",
    "            \n",
    "            BATCH_SIZE = 5\n",
    "#             if tok_length >= 1024:\n",
    "#                 BATCH_SIZE = 1\n",
    "            #print(BATCH_SIZE)\n",
    "            \n",
    "            with torch.no_grad():\n",
    "                past_key_values = list(model(input_ids=inputs[\"input_ids\"].to(DEVICE)).past_key_values)\n",
    "                \n",
    "                for idx0 in range(len(past_key_values)):\n",
    "                    past_key_values[idx0] = list(past_key_values[idx0])\n",
    "                    for idx1 in range(len(past_key_values[idx0])):\n",
    "                        past_key_values[idx0][idx1] = past_key_values[idx0][idx1].expand(BATCH_SIZE,-1,-1,-1)\n",
    "            del inputs\n",
    "                        \n",
    "        \n",
    "\n",
    "        if (idx+1) % BATCH_SIZE == 0 or idx == len(df) - 1:\n",
    "        #if True:\n",
    "\n",
    "            inputs = tokenizer(instructions, return_tensors='pt', add_special_tokens=False, truncation=True, padding=\"longest\")\n",
    "\n",
    "            with torch.no_grad():\n",
    "                out = model(input_ids=inputs[\"input_ids\"].to(DEVICE), past_key_values=past_key_values).logits\n",
    "                \n",
    "                for jjj in range(len(out)):\n",
    "                    att_idx = inputs[\"attention_mask\"].sum(dim=1)[jjj]-1\n",
    "                    pooled.append(out[jjj,att_idx,:].float().unsqueeze(0))\n",
    "                    \n",
    "            instructions = []\n",
    "            del out\n",
    "            del inputs\n",
    "         \n",
    "        if (idx+1) % 5 == 0:\n",
    "            with torch.no_grad():\n",
    "                pooled = torch.cat(pooled)\n",
    "                \n",
    "                if args.model_type == 2:\n",
    "                    new_poolings = pooled.reshape(1,-1)\n",
    "                    new_poolings = new_poolings.reshape(5, new_poolings.size(-1)//5)\n",
    "                    permutations = list(itertools.permutations(range(new_poolings.size(0))))\n",
    "                    tta_logits = torch.zeros((5,1)).to(new_poolings.device)\n",
    "                    for perm in permutations:\n",
    "                        permuted_poolings = new_poolings[torch.tensor(perm)]\n",
    "                        o = head(permuted_poolings.float().reshape(1,-1))\n",
    "                        tta_logits[torch.tensor(perm)] += o.reshape(-1,1)\n",
    "                    logits = tta_logits / len(permutations)\n",
    "                    logits = logits[:,0]\n",
    "                    logits = logits.detach().cpu().numpy()\n",
    "                else:\n",
    "                    new_poolings = []\n",
    "\n",
    "                    indexes = np.arange(0, 5)\n",
    "                    for jj in indexes:\n",
    "                        if args.model_type == 1:\n",
    "                            other_embeddings = pooled[[jjj for jjj in indexes if jjj != jj]]\n",
    "                            new_poolings.append(torch.cat([pooled[jj], torch.mean(other_embeddings, dim=0)]))\n",
    "                        else:\n",
    "                            new_poolings.append(pooled[jj])\n",
    "                    new_poolings = torch.stack(new_poolings)\n",
    "                    \n",
    "                    logits = head(new_poolings)\n",
    "                    logits = logits[:,0]\n",
    "                    logits = logits.detach().cpu().numpy()\n",
    "\n",
    "                for lg in logits:\n",
    "                    preds.append(lg)\n",
    "\n",
    "            del logits\n",
    "            pooled = []\n",
    "\n",
    "    print(inst)\n",
    "    np.save(f\"scores_{llm_backbone.split('/')[-1]}_{args.test_file[:-3]}\", preds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "677017f1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:10:42.853798Z",
     "iopub.status.busy": "2023-10-10T12:10:42.853163Z",
     "iopub.status.idle": "2023-10-10T12:10:42.858863Z",
     "shell.execute_reply": "2023-10-10T12:10:42.857857Z"
    },
    "papermill": {
     "duration": 0.044512,
     "end_time": "2023-10-10T12:10:42.860778",
     "exception": false,
     "start_time": "2023-10-10T12:10:42.816266",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing run.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile run.sh\n",
    "\n",
    "python inference.py --device \"auto\" --model_name \"/kaggle/input/teamhydrogen-cool-giraffe-prompt-2409\" --quantization 0 --model_type 1 --test_file \"test3.pq\"  &&\n",
    "python inference.py --device \"auto\" --model_name \"/kaggle/input/teamhydrogen-white-malamute-prompt-openorca-v2\" --quantization 0 --model_type 1 --test_file \"test.pq\"  &&\n",
    "python inference.py --device \"auto\" --model_name \"/kaggle/input/teamhydrogen-raspberry-beaver-1\" --quantization 0 --model_type 1 --test_file \"test5.pq\"  &&\n",
    "python inference.py --device \"auto\" --model_name \"/kaggle/input/teamhydrogen-cool-giraffe-prompt-0710\" --quantization 0 --model_type 1 --test_file \"test6.pq\"  &&\n",
    "\n",
    "wait \n",
    "echo \"All done\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ed3137f1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:10:42.930931Z",
     "iopub.status.busy": "2023-10-10T12:10:42.930630Z",
     "iopub.status.idle": "2023-10-10T12:40:36.786973Z",
     "shell.execute_reply": "2023-10-10T12:40:36.785745Z"
    },
    "papermill": {
     "duration": 1793.892936,
     "end_time": "2023-10-10T12:40:36.788974",
     "exception": false,
     "start_time": "2023-10-10T12:10:42.896038",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   id  ...                                             answer\r\n",
      "0   0  ...  RNA-templated transcription is the method of t...\r\n",
      "\r\n",
      "[1 rows x 5 columns]\r\n",
      "Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.\r\n",
      "Using pad_token, but it is not set yet.\r\n",
      "/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5\r\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']\r\n",
      "  warnings.warn(f\"unable to load libtensorflow_io_plugins.so: {e}\")\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tensorflow13GcsFileSystemE']\r\n",
      "  warnings.warn(f\"file system plugins are not loaded: {e}\")\r\n",
      "Loading checkpoint shards: 100%|██████████████████| 3/3 [01:57<00:00, 39.11s/it]\r\n",
      "64000\r\n",
      "  0%|                                          | 1/1500 [00:01<39:22,  1.58s/it]Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.\r\n",
      "100%|███████████████████████████████████████| 1500/1500 [04:49<00:00,  5.19it/s]\r\n",
      "Answer: T-duality explains the compactification of the extra dimensions in string/M-theory.\r\n",
      "###\r\n",
      "Is this answer correct? \r\n",
      "   id  ...                                             answer\r\n",
      "0   0  ...  RNA-templated transcription is the method of t...\r\n",
      "\r\n",
      "[1 rows x 5 columns]\r\n",
      "Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.\r\n",
      "Using pad_token, but it is not set yet.\r\n",
      "/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5\r\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']\r\n",
      "  warnings.warn(f\"unable to load libtensorflow_io_plugins.so: {e}\")\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tensorflow13GcsFileSystemE']\r\n",
      "  warnings.warn(f\"file system plugins are not loaded: {e}\")\r\n",
      "Loading checkpoint shards: 100%|██████████████████| 3/3 [02:07<00:00, 42.45s/it]\r\n",
      "64004\r\n",
      "  0%|                                          | 1/1500 [00:01<35:06,  1.41s/it]Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.\r\n",
      "100%|███████████████████████████████████████| 1500/1500 [05:17<00:00,  4.73it/s]\r\n",
      "Answer: T-duality explains the compactification of the extra dimensions in string/M-theory.\r\n",
      "###\r\n",
      "Is this answer correct? \r\n",
      "   id  ...                                             answer\r\n",
      "0   0  ...  RNA-templated transcription is the method of t...\r\n",
      "\r\n",
      "[1 rows x 5 columns]\r\n",
      "Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.\r\n",
      "/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5\r\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']\r\n",
      "  warnings.warn(f\"unable to load libtensorflow_io_plugins.so: {e}\")\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tensorflow13GcsFileSystemE']\r\n",
      "  warnings.warn(f\"file system plugins are not loaded: {e}\")\r\n",
      "Loading checkpoint shards: 100%|██████████████████| 3/3 [02:19<00:00, 46.39s/it]\r\n",
      "64000\r\n",
      "  0%|                                          | 1/1500 [00:01<33:56,  1.36s/it]Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.\r\n",
      "100%|███████████████████████████████████████| 1500/1500 [05:14<00:00,  4.77it/s]\r\n",
      "Answer: T-duality explains the compactification of the extra dimensions in string/M-theory.\r\n",
      "###\r\n",
      "Is this answer correct? \r\n",
      "   id  ...                                             answer\r\n",
      "0   0  ...  RNA-templated transcription is the method of t...\r\n",
      "\r\n",
      "[1 rows x 5 columns]\r\n",
      "Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.\r\n",
      "Using pad_token, but it is not set yet.\r\n",
      "/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5\r\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']\r\n",
      "  warnings.warn(f\"unable to load libtensorflow_io_plugins.so: {e}\")\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tensorflow13GcsFileSystemE']\r\n",
      "  warnings.warn(f\"file system plugins are not loaded: {e}\")\r\n",
      "Loading checkpoint shards: 100%|██████████████████| 3/3 [01:57<00:00, 39.03s/it]\r\n",
      "64000\r\n",
      "  0%|                                          | 1/1500 [00:01<40:44,  1.63s/it]Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.\r\n",
      "100%|███████████████████████████████████████| 1500/1500 [05:04<00:00,  4.93it/s]\r\n",
      "Answer: T-duality explains the compactification of the extra dimensions in string/M-theory.\r\n",
      "###\r\n",
      "Is this answer correct? \r\n",
      "All done\r\n"
     ]
    }
   ],
   "source": [
    "!sh run.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d2938856",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:40:36.965239Z",
     "iopub.status.busy": "2023-10-10T12:40:36.964904Z",
     "iopub.status.idle": "2023-10-10T12:40:37.147165Z",
     "shell.execute_reply": "2023-10-10T12:40:37.146268Z"
    },
    "papermill": {
     "duration": 0.270374,
     "end_time": "2023-10-10T12:40:37.149222",
     "exception": false,
     "start_time": "2023-10-10T12:40:36.878848",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "test = pd.read_parquet(\"test.pq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7a1030ac",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:40:37.402766Z",
     "iopub.status.busy": "2023-10-10T12:40:37.402414Z",
     "iopub.status.idle": "2023-10-10T12:40:37.406947Z",
     "shell.execute_reply": "2023-10-10T12:40:37.405921Z"
    },
    "papermill": {
     "duration": 0.092453,
     "end_time": "2023-10-10T12:40:37.408643",
     "exception": false,
     "start_time": "2023-10-10T12:40:37.316190",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "all_scores = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2663d2e0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:40:37.650960Z",
     "iopub.status.busy": "2023-10-10T12:40:37.650619Z",
     "iopub.status.idle": "2023-10-10T12:40:37.659404Z",
     "shell.execute_reply": "2023-10-10T12:40:37.658379Z"
    },
    "papermill": {
     "duration": 0.099963,
     "end_time": "2023-10-10T12:40:37.661174",
     "exception": false,
     "start_time": "2023-10-10T12:40:37.561211",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing inference_2.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile inference_2.py\n",
    "\n",
    "\n",
    "from transformers import AutoModelForCausalLM, AutoTokenizer\n",
    "import itertools\n",
    "import random\n",
    "import argparse\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from string import Template\n",
    "from pathlib import Path\n",
    "import time\n",
    "import torch\n",
    "from tqdm.auto import tqdm\n",
    "import gc\n",
    "import itertools\n",
    "\n",
    "#os.environ[\"TOKENIZERS_PARALLELISM\"] = \"false\"\n",
    "\n",
    "def precision_at_k(r, k):\n",
    "    \"\"\"Precision at k\"\"\"\n",
    "    assert k <= len(r)\n",
    "    assert k != 0\n",
    "    return sum(int(x) for x in r[:k]) / k\n",
    "\n",
    "def MAP_at_3(predictions, true_items):\n",
    "    \"\"\"Score is mean average precision at 3\"\"\"\n",
    "    U = len(predictions)\n",
    "    map_at_3 = 0.0\n",
    "    for u in range(U):\n",
    "        user_preds = predictions[u]\n",
    "        user_true = true_items[u]\n",
    "        user_results = [1 if item == user_true else 0 for item in user_preds]\n",
    "        for k in range(min(len(user_preds), 3)):\n",
    "            map_at_3 += precision_at_k(user_results, k+1) * user_results[k]\n",
    "    return map_at_3 / U\n",
    "\n",
    "def longest_common_prefix(strs):\n",
    "    if not strs:\n",
    "        return \"\"\n",
    "\n",
    "    shortest = min(strs,key=len)\n",
    "\n",
    "    for i, char in enumerate(shortest):\n",
    "        for other in strs:\n",
    "            if other[i] != char:\n",
    "                return shortest[:i]\n",
    "\n",
    "    return shortest \n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    \n",
    "    ap = argparse.ArgumentParser()\n",
    "    ap.add_argument('--device', type=str, required=True)\n",
    "    ap.add_argument('--model_name', type=str, required=True)\n",
    "    ap.add_argument('--quantization', type=int, required=True)\n",
    "    ap.add_argument('--model_type', type=int, required=True)\n",
    "    ap.add_argument('--test_file', type=str, required=True)\n",
    "    args = ap.parse_args()\n",
    "    \n",
    "    if args.device == \"auto\":\n",
    "        DEVICE_MAP = \"auto\"\n",
    "        DEVICE = \"cuda:0\"\n",
    "    else:\n",
    "        DEVICE_MAP = {\"\":args.device}\n",
    "        DEVICE = args.device\n",
    "    \n",
    "    llm_backbone = args.model_name\n",
    "    \n",
    "    test = pd.read_parquet(args.test_file).reset_index(drop=True)\n",
    "\n",
    "\n",
    "    new_obs = []\n",
    "    for idx, row in test.iterrows():\n",
    "        for opt in \"ABCDE\":\n",
    "            new_obs.append((row[\"id\"], row[\"context\"], row[\"prompt\"], row[opt]))\n",
    "    df = pd.DataFrame(new_obs, columns=[\"id\", \"context\", \"question\", \"answer\"])\n",
    "\n",
    "    \n",
    "    print(df.head(1))\n",
    "    \n",
    "\n",
    "    tokenizer = AutoTokenizer.from_pretrained(\n",
    "        llm_backbone,\n",
    "        use_fast=True,\n",
    "        trust_remote_code=True,\n",
    "        padding_side=\"right\",\n",
    "        truncation_side=\"left\"\n",
    "    )\n",
    "    \n",
    "    if tokenizer.pad_token is None:\n",
    "        if tokenizer.unk_token is not None:\n",
    "            tokenizer.pad_token = tokenizer.unk_token\n",
    "        else:\n",
    "            tokenizer.pad_token = tokenizer.eos_token\n",
    "    \n",
    "    from transformers import BitsAndBytesConfig\n",
    "    \n",
    "    \n",
    "#     quantization_config = BitsAndBytesConfig(\n",
    "#        load_in_4bit=True,\n",
    "#        bnb_4bit_quant_type=\"nf4\",\n",
    "#        bnb_4bit_use_double_quant=False,\n",
    "#        bnb_4bit_compute_dtype=torch.float16\n",
    "#     )\n",
    "\n",
    "    if args.quantization == 0:\n",
    "        quantization_config = None\n",
    "\n",
    "    elif args.quantization == 1:\n",
    "        quantization_config = BitsAndBytesConfig(\n",
    "            load_in_8bit=True,\n",
    "            llm_int8_threshold=0.0,\n",
    "        )\n",
    "\n",
    "\n",
    "    model = AutoModelForCausalLM.from_pretrained(\n",
    "        llm_backbone,\n",
    "        torch_dtype=torch.float16,\n",
    "        quantization_config=quantization_config,\n",
    "        device_map=DEVICE_MAP,\n",
    "        low_cpu_mem_usage=True,\n",
    "        trust_remote_code=True\n",
    "    ).eval()\n",
    "\n",
    "    head_weights = torch.load(llm_backbone + \"/head.pth\", map_location=\"cpu\")\n",
    "    hidden_size = head_weights.shape[1]\n",
    "    print(hidden_size)\n",
    "    # hidden_size = 32000\n",
    "    head = torch.nn.Linear(hidden_size, 1, bias=False)\n",
    "\n",
    "    head.weight.data = head_weights\n",
    "\n",
    "    head.to(DEVICE).eval()\n",
    "\n",
    "    model.config.pad_token_id = tokenizer.pad_token_id\n",
    "    \n",
    "    gc.collect()\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "    import itertools\n",
    "    import random\n",
    "    maps = []\n",
    "\n",
    "    progress_bar = tqdm(df.iterrows(), total=len(df))\n",
    "\n",
    "    preds = []\n",
    "    instructions = []\n",
    "    pooled = []\n",
    "    past_key_values = None\n",
    "    for idx,row in progress_bar:\n",
    "        \n",
    "        inst = f\"{row['answer']}\"\n",
    "        #inst = \"a b c d\"*10_000\n",
    "        instructions.append(inst)\n",
    "        \n",
    "        if idx % 5 == 0:\n",
    "            \n",
    "            if past_key_values is not None:\n",
    "                del past_key_values\n",
    "            \n",
    "            preprompt = f\"{row['context']}\\n###\\n{row['question']}\\n###\\n\"\n",
    "            #preprompt = \"a b c d\"*10_000\n",
    "            inputs = tokenizer(preprompt, return_tensors='pt', add_special_tokens=False, truncation=True, padding=\"longest\", max_length=1024)\n",
    "            \n",
    "            tok_length = inputs[\"input_ids\"].shape[1] + tokenizer(instructions, return_tensors='pt', add_special_tokens=False, truncation=True, padding=\"longest\", max_length=1024)[\"input_ids\"].shape[1]\n",
    "            \n",
    "            BATCH_SIZE = 5\n",
    "#             if tok_length >= 1024:\n",
    "#                 BATCH_SIZE = 1\n",
    "            #print(BATCH_SIZE)\n",
    "            \n",
    "            with torch.no_grad():\n",
    "                past_key_values = list(model(input_ids=inputs[\"input_ids\"].to(DEVICE)).past_key_values)\n",
    "                \n",
    "                for idx0 in range(len(past_key_values)):\n",
    "                    past_key_values[idx0] = list(past_key_values[idx0])\n",
    "                    for idx1 in range(len(past_key_values[idx0])):\n",
    "                        past_key_values[idx0][idx1] = past_key_values[idx0][idx1].expand(BATCH_SIZE,-1,-1,-1)\n",
    "            del inputs\n",
    "                        \n",
    "        \n",
    "\n",
    "        if (idx+1) % BATCH_SIZE == 0 or idx == len(df) - 1:\n",
    "        #if True:\n",
    "\n",
    "            inputs = tokenizer(instructions, return_tensors='pt', add_special_tokens=False, truncation=True, padding=\"longest\")\n",
    "\n",
    "            with torch.no_grad():\n",
    "                out = model(input_ids=inputs[\"input_ids\"].to(DEVICE), past_key_values=past_key_values).logits\n",
    "                \n",
    "                for jjj in range(len(out)):\n",
    "                    att_idx = inputs[\"attention_mask\"].sum(dim=1)[jjj]-1\n",
    "                    pooled.append(out[jjj,att_idx,:].float().unsqueeze(0))\n",
    "                    \n",
    "            instructions = []\n",
    "            del out\n",
    "            del inputs\n",
    "         \n",
    "        if (idx+1) % 5 == 0:\n",
    "            with torch.no_grad():\n",
    "                pooled = torch.cat(pooled)\n",
    "                \n",
    "                if args.model_type == 2:\n",
    "                    new_poolings = pooled.reshape(1,-1)\n",
    "                    new_poolings = new_poolings.reshape(5, new_poolings.size(-1)//5)\n",
    "                    permutations = list(itertools.permutations(range(new_poolings.size(0))))\n",
    "                    tta_logits = torch.zeros((5,1)).to(new_poolings.device)\n",
    "                    for perm in permutations:\n",
    "                        permuted_poolings = new_poolings[torch.tensor(perm)]\n",
    "                        o = head(permuted_poolings.float().reshape(1,-1))\n",
    "                        tta_logits[torch.tensor(perm)] += o.reshape(-1,1)\n",
    "                    logits = tta_logits / len(permutations)\n",
    "                    logits = logits[:,0]\n",
    "                    logits = logits.detach().cpu().numpy()\n",
    "                else:\n",
    "                    new_poolings = []\n",
    "\n",
    "                    indexes = np.arange(0, 5)\n",
    "                    for jj in indexes:\n",
    "                        if args.model_type == 1:\n",
    "                            other_embeddings = pooled[[jjj for jjj in indexes if jjj != jj]]\n",
    "                            new_poolings.append(torch.cat([pooled[jj], torch.mean(other_embeddings, dim=0)]))\n",
    "                        else:\n",
    "                            new_poolings.append(pooled[jj])\n",
    "                    new_poolings = torch.stack(new_poolings)\n",
    "                    \n",
    "                    logits = head(new_poolings)\n",
    "                    logits = logits[:,0]\n",
    "                    logits = logits.detach().cpu().numpy()\n",
    "\n",
    "                for lg in logits:\n",
    "                    preds.append(lg)\n",
    "\n",
    "            del logits\n",
    "            pooled = []\n",
    "\n",
    "    print(inst)\n",
    "    np.save(f\"scores_{llm_backbone.split('/')[-1]}_{args.test_file[:-3]}\", preds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a482d5c6",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:40:37.841229Z",
     "iopub.status.busy": "2023-10-10T12:40:37.840940Z",
     "iopub.status.idle": "2023-10-10T12:40:37.846285Z",
     "shell.execute_reply": "2023-10-10T12:40:37.845313Z"
    },
    "papermill": {
     "duration": 0.096769,
     "end_time": "2023-10-10T12:40:37.847976",
     "exception": false,
     "start_time": "2023-10-10T12:40:37.751207",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting run.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile run.sh\n",
    "\n",
    "python inference_2.py --device \"auto\" --model_name \"/kaggle/input/teamhydrogen-nostalgic-mammoth-avg\" --quantization 0 --model_type 1 --test_file \"test4.pq\"  &&\n",
    "\n",
    "wait \n",
    "echo \"All done\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "90baaf2e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:40:38.019263Z",
     "iopub.status.busy": "2023-10-10T12:40:38.018346Z",
     "iopub.status.idle": "2023-10-10T12:45:46.022855Z",
     "shell.execute_reply": "2023-10-10T12:45:46.021715Z"
    },
    "papermill": {
     "duration": 308.093072,
     "end_time": "2023-10-10T12:45:46.024965",
     "exception": false,
     "start_time": "2023-10-10T12:40:37.931893",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   id  ...                                             answer\r\n",
      "0   0  ...  RNA-templated transcription is the method of t...\r\n",
      "\r\n",
      "[1 rows x 4 columns]\r\n",
      "/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5\r\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']\r\n",
      "  warnings.warn(f\"unable to load libtensorflow_io_plugins.so: {e}\")\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tensorflow13GcsFileSystemE']\r\n",
      "  warnings.warn(f\"file system plugins are not loaded: {e}\")\r\n",
      "Loading checkpoint shards: 100%|██████████████████| 3/3 [02:07<00:00, 42.57s/it]\r\n",
      "64000\r\n",
      "  0%|                                          | 1/1500 [00:01<37:50,  1.51s/it]Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.\r\n",
      "100%|███████████████████████████████████████| 1500/1500 [02:43<00:00,  9.15it/s]\r\n",
      "T-duality explains the compactification of the extra dimensions in string/M-theory.\r\n",
      "All done\r\n"
     ]
    }
   ],
   "source": [
    "!sh run.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "efa4a84b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:45:46.223625Z",
     "iopub.status.busy": "2023-10-10T12:45:46.223301Z",
     "iopub.status.idle": "2023-10-10T12:45:46.230156Z",
     "shell.execute_reply": "2023-10-10T12:45:46.229023Z"
    },
    "papermill": {
     "duration": 0.107358,
     "end_time": "2023-10-10T12:45:46.231844",
     "exception": false,
     "start_time": "2023-10-10T12:45:46.124486",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting run.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile run.sh\n",
    "\n",
    "python inference_2.py --device \"auto\" --model_name \"/kaggle/input/teamhydrogen-wealthy-ant-llama2-13b-1909\" --quantization 1 --model_type 0 --test_file \"test2.pq\"  &&\n",
    "\n",
    "wait \n",
    "echo \"All done\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "003f2a77",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:45:46.430981Z",
     "iopub.status.busy": "2023-10-10T12:45:46.430119Z",
     "iopub.status.idle": "2023-10-10T12:55:07.492369Z",
     "shell.execute_reply": "2023-10-10T12:55:07.490736Z"
    },
    "papermill": {
     "duration": 561.168485,
     "end_time": "2023-10-10T12:55:07.496869",
     "exception": false,
     "start_time": "2023-10-10T12:45:46.328384",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   id  ...                                             answer\r\n",
      "0   0  ...  RNA-templated transcription is the method of t...\r\n",
      "\r\n",
      "[1 rows x 4 columns]\r\n",
      "/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5\r\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']\r\n",
      "  warnings.warn(f\"unable to load libtensorflow_io_plugins.so: {e}\")\r\n",
      "/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']\r\n",
      "caused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tensorflow13GcsFileSystemE']\r\n",
      "  warnings.warn(f\"file system plugins are not loaded: {e}\")\r\n",
      "Loading checkpoint shards: 100%|██████████████████| 3/3 [02:18<00:00, 46.27s/it]\r\n",
      "32000\r\n",
      "  0%|                                          | 1/1500 [00:02<57:37,  2.31s/it]Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.\r\n",
      "100%|███████████████████████████████████████| 1500/1500 [06:43<00:00,  3.72it/s]\r\n",
      "T-duality explains the compactification of the extra dimensions in string/M-theory.\r\n",
      "All done\r\n"
     ]
    }
   ],
   "source": [
    "!sh run.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8991c6d2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:55:07.927708Z",
     "iopub.status.busy": "2023-10-10T12:55:07.927366Z",
     "iopub.status.idle": "2023-10-10T12:55:08.173498Z",
     "shell.execute_reply": "2023-10-10T12:55:08.172361Z"
    },
    "papermill": {
     "duration": 0.375269,
     "end_time": "2023-10-10T12:55:08.176687",
     "exception": false,
     "start_time": "2023-10-10T12:55:07.801418",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scores_teamhydrogen-raspberry-beaver-1_test5.npy\n",
      "0.2\n",
      "scores_teamhydrogen-wealthy-ant-llama2-13b-1909_test2.npy\n",
      "0.2\n",
      "scores_teamhydrogen-white-malamute-prompt-openorca-v2_test.npy\n",
      "0.2\n",
      "scores_teamhydrogen-cool-giraffe-prompt-2409_test3.npy\n",
      "0.2\n",
      "scores_teamhydrogen-cool-giraffe-prompt-0710_test6.npy\n",
      "0.2\n",
      "scores_teamhydrogen-nostalgic-mammoth-avg_test4.npy\n",
      "0.2\n"
     ]
    }
   ],
   "source": [
    "import glob\n",
    "from scipy.special import softmax\n",
    "\n",
    "curr_scores = []\n",
    "\n",
    "for f in glob.glob(\"scores_*.npy\"):\n",
    "    print(f)\n",
    "    a = np.array(np.load(f))\n",
    "    a = softmax(a.reshape(-1,5), axis=1)\n",
    "    a = a.flatten()\n",
    "    print(a.mean(axis=0))\n",
    "\n",
    "    curr_scores.append(a)\n",
    "    os.remove(f)\n",
    "curr_scores = np.array(curr_scores)\n",
    "all_scores.append(np.nanmean(curr_scores, axis=0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1b1fc00b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:55:08.460587Z",
     "iopub.status.busy": "2023-10-10T12:55:08.460192Z",
     "iopub.status.idle": "2023-10-10T12:55:08.465559Z",
     "shell.execute_reply": "2023-10-10T12:55:08.464583Z"
    },
    "papermill": {
     "duration": 0.133985,
     "end_time": "2023-10-10T12:55:08.467408",
     "exception": false,
     "start_time": "2023-10-10T12:55:08.333423",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "all_scores = np.nanmean(all_scores, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "993a7470",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:55:08.728517Z",
     "iopub.status.busy": "2023-10-10T12:55:08.728203Z",
     "iopub.status.idle": "2023-10-10T12:55:08.734284Z",
     "shell.execute_reply": "2023-10-10T12:55:08.733303Z"
    },
    "papermill": {
     "duration": 0.132812,
     "end_time": "2023-10-10T12:55:08.736012",
     "exception": false,
     "start_time": "2023-10-10T12:55:08.603200",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1500,)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_scores.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "9fc0490c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:55:08.988852Z",
     "iopub.status.busy": "2023-10-10T12:55:08.988540Z",
     "iopub.status.idle": "2023-10-10T12:55:08.994260Z",
     "shell.execute_reply": "2023-10-10T12:55:08.993352Z"
    },
    "papermill": {
     "duration": 0.132904,
     "end_time": "2023-10-10T12:55:08.995987",
     "exception": false,
     "start_time": "2023-10-10T12:55:08.863083",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1500,)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preds = all_scores\n",
    "preds.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "6d944bf5",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:55:09.255926Z",
     "iopub.status.busy": "2023-10-10T12:55:09.255596Z",
     "iopub.status.idle": "2023-10-10T12:55:09.261017Z",
     "shell.execute_reply": "2023-10-10T12:55:09.260130Z"
    },
    "papermill": {
     "duration": 0.133803,
     "end_time": "2023-10-10T12:55:09.262669",
     "exception": false,
     "start_time": "2023-10-10T12:55:09.128866",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1500,)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preds = np.array(preds)\n",
    "preds.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d0986918",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:55:09.512648Z",
     "iopub.status.busy": "2023-10-10T12:55:09.512358Z",
     "iopub.status.idle": "2023-10-10T12:55:09.525524Z",
     "shell.execute_reply": "2023-10-10T12:55:09.524529Z"
    },
    "papermill": {
     "duration": 0.14041,
     "end_time": "2023-10-10T12:55:09.527184",
     "exception": false,
     "start_time": "2023-10-10T12:55:09.386774",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "300"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We'll create a dictionary to convert option names (A, B, C, D, E) into indices and back again\n",
    "options = 'ABCDE'\n",
    "indices = list(range(5))\n",
    "\n",
    "option_to_index = {option: index for option, index in zip(options, indices)}\n",
    "index_to_option = {index: option for option, index in zip(options, indices)}\n",
    "post_proc_preds = []\n",
    "\n",
    "for gr in range(0, len(preds), 5):\n",
    "    pr = preds[gr : gr + 5]\n",
    "    pr = np.argsort(-pr)[:3]\n",
    "    post_proc_preds.append(\" \".join([index_to_option[x] for x in pr]))\n",
    "len(post_proc_preds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c9575578",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:55:09.827256Z",
     "iopub.status.busy": "2023-10-10T12:55:09.826919Z",
     "iopub.status.idle": "2023-10-10T12:55:09.832757Z",
     "shell.execute_reply": "2023-10-10T12:55:09.831696Z"
    },
    "papermill": {
     "duration": 0.183645,
     "end_time": "2023-10-10T12:55:09.834502",
     "exception": false,
     "start_time": "2023-10-10T12:55:09.650857",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "test['prediction'] = post_proc_preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "7dedd65e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:55:10.089523Z",
     "iopub.status.busy": "2023-10-10T12:55:10.088704Z",
     "iopub.status.idle": "2023-10-10T12:55:10.122880Z",
     "shell.execute_reply": "2023-10-10T12:55:10.119865Z"
    },
    "papermill": {
     "duration": 0.16209,
     "end_time": "2023-10-10T12:55:10.125428",
     "exception": false,
     "start_time": "2023-10-10T12:55:09.963338",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.5 D ['A', 'D', 'B']\n",
      "1.0 D ['D', 'C', 'B']\n",
      "0.0 A ['C', 'B', 'E']\n",
      "0.0 B ['C', 'E', 'A']\n",
      "1.0 A ['A', 'D', 'C']\n",
      "0.5 B ['D', 'B', 'E']\n",
      "1.0 B ['B', 'D', 'E']\n",
      "0.0 A ['D', 'C', 'B']\n",
      "0.0 E ['A', 'C', 'D']\n",
      "0.5 A ['D', 'A', 'C']\n",
      "1.0 B ['B', 'D', 'A']\n",
      "0.5 D ['C', 'D', 'E']\n",
      "1.0 D ['D', 'A', 'B']\n",
      "1.0 D ['D', 'C', 'B']\n",
      "1.0 A ['A', 'B', 'E']\n",
      "1.0 D ['D', 'B', 'A']\n",
      "1.0 D ['D', 'E', 'C']\n",
      "0.5 C ['D', 'C', 'E']\n",
      "1.0 A ['A', 'B', 'C']\n",
      "1.0 B ['B', 'D', 'E']\n",
      "0.3333333333333333 A ['E', 'B', 'A']\n",
      "1.0 D ['D', 'C', 'B']\n",
      "0.5 D ['A', 'D', 'E']\n",
      "0.5 A ['D', 'A', 'E']\n",
      "1.0 B ['B', 'D', 'A']\n",
      "0.5 C ['E', 'C', 'A']\n",
      "0.5 C ['A', 'C', 'D']\n",
      "0.3333333333333333 D ['C', 'B', 'D']\n",
      "1.0 E ['E', 'D', 'C']\n",
      "0.5 D ['B', 'D', 'A']\n",
      "0.3333333333333333 E ['C', 'B', 'E']\n",
      "1.0 E ['E', 'B', 'D']\n",
      "0.0 B ['C', 'A', 'E']\n",
      "1.0 D ['D', 'C', 'B']\n",
      "1.0 D ['D', 'A', 'C']\n",
      "0.0 B ['D', 'A', 'C']\n",
      "1.0 D ['D', 'B', 'A']\n",
      "1.0 C ['C', 'A', 'D']\n",
      "1.0 E ['E', 'C', 'D']\n",
      "1.0 C ['C', 'B', 'A']\n",
      "0.3333333333333333 B ['E', 'D', 'B']\n",
      "1.0 D ['D', 'B', 'E']\n",
      "1.0 A ['A', 'B', 'D']\n",
      "1.0 D ['D', 'B', 'E']\n",
      "0.0 E ['D', 'C', 'B']\n",
      "0.0 B ['D', 'E', 'C']\n",
      "0.0 B ['C', 'E', 'D']\n",
      "1.0 E ['E', 'A', 'B']\n",
      "0.3333333333333333 A ['E', 'D', 'A']\n",
      "1.0 C ['C', 'A', 'E']\n",
      "0.0 D ['C', 'B', 'A']\n",
      "1.0 E ['E', 'D', 'C']\n",
      "1.0 C ['C', 'D', 'E']\n",
      "1.0 E ['E', 'D', 'B']\n",
      "1.0 B ['B', 'C', 'E']\n",
      "0.3333333333333333 C ['D', 'B', 'C']\n",
      "0.5 A ['D', 'A', 'E']\n",
      "0.0 D ['C', 'B', 'E']\n",
      "1.0 D ['D', 'A', 'C']\n",
      "1.0 B ['B', 'E', 'D']\n",
      "1.0 D ['D', 'A', 'E']\n",
      "1.0 B ['B', 'D', 'A']\n",
      "1.0 D ['D', 'C', 'B']\n",
      "1.0 B ['B', 'A', 'E']\n",
      "1.0 C ['C', 'D', 'E']\n",
      "1.0 A ['A', 'C', 'D']\n",
      "1.0 A ['A', 'C', 'E']\n",
      "0.5 D ['B', 'D', 'C']\n",
      "1.0 D ['D', 'A', 'B']\n",
      "0.5 B ['E', 'B', 'A']\n",
      "0.5 A ['D', 'A', 'C']\n",
      "1.0 A ['A', 'C', 'B']\n",
      "0.5 A ['E', 'A', 'D']\n",
      "0.0 E ['B', 'D', 'C']\n",
      "0.5 A ['E', 'A', 'B']\n",
      "0.0 A ['D', 'C', 'E']\n",
      "0.5 A ['C', 'A', 'B']\n",
      "1.0 B ['B', 'C', 'E']\n",
      "1.0 A ['A', 'D', 'E']\n",
      "1.0 B ['B', 'D', 'E']\n",
      "0.5 A ['B', 'A', 'E']\n",
      "0.5 A ['C', 'A', 'B']\n",
      "1.0 B ['B', 'C', 'E']\n",
      "1.0 B ['B', 'E', 'C']\n",
      "0.0 C ['B', 'E', 'D']\n",
      "0.3333333333333333 A ['E', 'D', 'A']\n",
      "1.0 E ['E', 'C', 'A']\n",
      "1.0 A ['A', 'E', 'D']\n",
      "1.0 E ['E', 'A', 'C']\n",
      "0.5 B ['C', 'B', 'A']\n",
      "1.0 E ['E', 'B', 'D']\n",
      "0.5 B ['C', 'B', 'A']\n",
      "1.0 C ['C', 'D', 'A']\n",
      "0.5 E ['B', 'E', 'A']\n",
      "1.0 A ['A', 'E', 'D']\n",
      "0.0 C ['D', 'E', 'A']\n",
      "0.5 C ['D', 'C', 'B']\n",
      "0.5 E ['D', 'E', 'B']\n",
      "0.5 A ['B', 'A', 'E']\n",
      "1.0 B ['B', 'C', 'E']\n",
      "1.0 B ['B', 'A', 'C']\n",
      "0.0 C ['B', 'E', 'D']\n",
      "1.0 E ['E', 'C', 'A']\n",
      "1.0 A ['A', 'B', 'D']\n",
      "1.0 E ['E', 'A', 'B']\n",
      "0.0 A ['C', 'B', 'D']\n",
      "1.0 A ['A', 'D', 'E']\n",
      "0.5 D ['C', 'D', 'E']\n",
      "1.0 E ['E', 'C', 'D']\n",
      "0.5 C ['D', 'C', 'A']\n",
      "1.0 D ['D', 'B', 'A']\n",
      "0.3333333333333333 C ['B', 'E', 'C']\n",
      "1.0 B ['B', 'C', 'E']\n",
      "1.0 C ['C', 'D', 'E']\n",
      "1.0 B ['B', 'C', 'E']\n",
      "1.0 A ['A', 'B', 'C']\n",
      "1.0 C ['C', 'B', 'E']\n",
      "0.5 A ['C', 'A', 'D']\n",
      "1.0 C ['C', 'E', 'A']\n",
      "1.0 E ['E', 'C', 'B']\n",
      "1.0 B ['B', 'A', 'E']\n",
      "0.5 C ['B', 'C', 'A']\n",
      "1.0 B ['B', 'E', 'C']\n",
      "1.0 A ['A', 'E', 'D']\n",
      "1.0 A ['A', 'E', 'D']\n",
      "1.0 C ['C', 'D', 'E']\n",
      "1.0 E ['E', 'D', 'B']\n",
      "0.5 D ['B', 'D', 'E']\n",
      "0.5 A ['C', 'A', 'B']\n",
      "1.0 E ['E', 'A', 'D']\n",
      "1.0 A ['A', 'B', 'E']\n",
      "1.0 D ['D', 'C', 'A']\n",
      "1.0 A ['A', 'C', 'B']\n",
      "1.0 E ['E', 'B', 'D']\n",
      "1.0 D ['D', 'C', 'A']\n",
      "0.5 A ['B', 'A', 'C']\n",
      "1.0 D ['D', 'C', 'B']\n",
      "1.0 D ['D', 'C', 'E']\n",
      "0.0 C ['A', 'E', 'D']\n",
      "1.0 B ['B', 'D', 'E']\n",
      "1.0 A ['A', 'D', 'E']\n",
      "1.0 D ['D', 'C', 'E']\n",
      "1.0 B ['B', 'A', 'E']\n",
      "0.0 B ['A', 'C', 'E']\n",
      "1.0 A ['A', 'D', 'C']\n",
      "0.5 C ['E', 'C', 'D']\n",
      "1.0 D ['D', 'E', 'A']\n",
      "1.0 B ['B', 'E', 'C']\n",
      "0.5 B ['D', 'B', 'A']\n",
      "1.0 D ['D', 'C', 'A']\n",
      "0.5 C ['E', 'C', 'B']\n",
      "1.0 A ['A', 'C', 'D']\n",
      "0.5 C ['D', 'C', 'A']\n",
      "0.0 C ['B', 'D', 'A']\n",
      "1.0 C ['C', 'D', 'B']\n",
      "1.0 B ['B', 'A', 'E']\n",
      "1.0 C ['C', 'D', 'E']\n",
      "1.0 A ['A', 'E', 'C']\n",
      "0.5 C ['B', 'C', 'D']\n",
      "1.0 A ['A', 'E', 'C']\n",
      "0.5 C ['B', 'C', 'A']\n",
      "1.0 D ['D', 'A', 'B']\n",
      "0.5 E ['B', 'E', 'A']\n",
      "0.0 B ['A', 'D', 'C']\n",
      "1.0 B ['B', 'D', 'C']\n",
      "1.0 A ['A', 'B', 'C']\n",
      "1.0 A ['A', 'E', 'B']\n",
      "1.0 B ['B', 'A', 'E']\n",
      "1.0 B ['B', 'D', 'E']\n",
      "1.0 E ['E', 'C', 'D']\n",
      "1.0 D ['D', 'B', 'C']\n",
      "1.0 E ['E', 'A', 'C']\n",
      "1.0 A ['A', 'E', 'B']\n",
      "1.0 D ['D', 'E', 'B']\n",
      "1.0 D ['D', 'C', 'A']\n",
      "0.5 D ['E', 'D', 'B']\n",
      "1.0 D ['D', 'C', 'A']\n",
      "1.0 B ['B', 'E', 'D']\n",
      "1.0 A ['A', 'C', 'E']\n",
      "1.0 D ['D', 'A', 'B']\n",
      "1.0 C ['C', 'A', 'E']\n",
      "1.0 D ['D', 'B', 'A']\n",
      "0.0 E ['B', 'C', 'A']\n",
      "1.0 D ['D', 'B', 'E']\n",
      "1.0 C ['C', 'D', 'B']\n",
      "1.0 E ['E', 'A', 'C']\n",
      "1.0 D ['D', 'B', 'C']\n",
      "1.0 C ['C', 'D', 'A']\n",
      "1.0 A ['A', 'C', 'E']\n",
      "0.3333333333333333 B ['A', 'D', 'B']\n",
      "1.0 D ['D', 'A', 'B']\n",
      "1.0 E ['E', 'C', 'A']\n",
      "1.0 A ['A', 'C', 'B']\n",
      "0.3333333333333333 A ['C', 'E', 'A']\n",
      "0.0 B ['A', 'E', 'C']\n",
      "0.3333333333333333 C ['E', 'B', 'C']\n",
      "1.0 A ['A', 'D', 'C']\n",
      "1.0 E ['E', 'B', 'A']\n",
      "1.0 C ['C', 'B', 'E']\n",
      "1.0 B ['B', 'C', 'A']\n",
      "1.0 C ['C', 'B', 'D']\n",
      "1.0 A ['A', 'C', 'E']\n",
      "1.0 B ['B', 'C', 'D']\n",
      "0.0 B ['E', 'D', 'A']\n",
      "1.0 A ['A', 'C', 'D']\n",
      "1.0 E ['E', 'D', 'A']\n",
      "0.5 C ['B', 'C', 'E']\n",
      "0.3333333333333333 E ['B', 'C', 'E']\n",
      "0.3333333333333333 E ['B', 'C', 'E']\n",
      "1.0 D ['D', 'E', 'B']\n",
      "1.0 C ['C', 'E', 'B']\n",
      "1.0 B ['B', 'E', 'D']\n",
      "1.0 E ['E', 'D', 'B']\n",
      "1.0 E ['E', 'A', 'D']\n",
      "1.0 E ['E', 'B', 'C']\n",
      "0.3333333333333333 E ['A', 'B', 'E']\n",
      "1.0 C ['C', 'E', 'D']\n",
      "0.5 A ['D', 'A', 'B']\n",
      "1.0 B ['B', 'E', 'D']\n",
      "0.0 D ['A', 'E', 'B']\n",
      "1.0 C ['C', 'E', 'D']\n",
      "1.0 E ['E', 'C', 'A']\n",
      "0.5 D ['C', 'D', 'A']\n",
      "1.0 A ['A', 'B', 'D']\n",
      "1.0 A ['A', 'E', 'B']\n",
      "1.0 A ['A', 'C', 'B']\n",
      "0.3333333333333333 A ['C', 'E', 'A']\n",
      "1.0 D ['D', 'B', 'C']\n",
      "1.0 D ['D', 'A', 'B']\n",
      "0.5 A ['C', 'A', 'E']\n",
      "1.0 A ['A', 'D', 'E']\n",
      "0.0 A ['C', 'E', 'B']\n",
      "1.0 A ['A', 'C', 'D']\n",
      "1.0 D ['D', 'E', 'A']\n",
      "1.0 D ['D', 'C', 'B']\n",
      "1.0 D ['D', 'B', 'A']\n",
      "0.3333333333333333 C ['A', 'D', 'C']\n",
      "1.0 E ['E', 'B', 'A']\n",
      "0.5 A ['D', 'A', 'C']\n",
      "1.0 B ['B', 'E', 'A']\n",
      "1.0 D ['D', 'B', 'C']\n",
      "1.0 C ['C', 'D', 'E']\n",
      "1.0 C ['C', 'A', 'B']\n",
      "1.0 A ['A', 'C', 'B']\n",
      "1.0 D ['D', 'C', 'A']\n",
      "1.0 D ['D', 'B', 'E']\n",
      "0.5 E ['D', 'E', 'B']\n",
      "0.3333333333333333 E ['C', 'A', 'E']\n",
      "1.0 B ['B', 'A', 'D']\n",
      "1.0 D ['D', 'C', 'E']\n",
      "1.0 E ['E', 'B', 'D']\n",
      "0.0 C ['A', 'E', 'B']\n",
      "1.0 C ['C', 'E', 'B']\n",
      "1.0 E ['E', 'C', 'A']\n",
      "1.0 A ['A', 'B', 'D']\n",
      "1.0 D ['D', 'B', 'E']\n",
      "0.0 B ['C', 'E', 'D']\n",
      "1.0 D ['D', 'A', 'C']\n",
      "0.3333333333333333 E ['D', 'C', 'E']\n",
      "1.0 B ['B', 'D', 'C']\n",
      "0.0 E ['B', 'A', 'D']\n",
      "1.0 C ['C', 'E', 'D']\n",
      "0.5 B ['C', 'B', 'E']\n",
      "1.0 C ['C', 'D', 'A']\n",
      "1.0 C ['C', 'D', 'E']\n",
      "1.0 A ['A', 'B', 'D']\n",
      "0.5 C ['E', 'C', 'A']\n",
      "1.0 B ['B', 'C', 'E']\n",
      "0.0 C ['B', 'E', 'D']\n",
      "1.0 B ['B', 'E', 'A']\n",
      "0.5 C ['E', 'C', 'D']\n",
      "1.0 C ['C', 'B', 'A']\n",
      "1.0 D ['D', 'E', 'C']\n",
      "1.0 A ['A', 'D', 'C']\n",
      "1.0 A ['A', 'E', 'B']\n",
      "1.0 A ['A', 'B', 'E']\n",
      "1.0 E ['E', 'B', 'A']\n",
      "1.0 D ['D', 'C', 'B']\n",
      "0.0 C ['A', 'D', 'E']\n",
      "0.5 A ['D', 'A', 'B']\n",
      "1.0 B ['B', 'C', 'D']\n",
      "1.0 E ['E', 'C', 'A']\n",
      "1.0 B ['B', 'A', 'D']\n",
      "1.0 C ['C', 'E', 'A']\n",
      "1.0 D ['D', 'B', 'E']\n",
      "1.0 D ['D', 'C', 'A']\n",
      "1.0 D ['D', 'A', 'E']\n",
      "0.0 C ['D', 'A', 'E']\n",
      "1.0 D ['D', 'B', 'C']\n",
      "1.0 B ['B', 'C', 'D']\n",
      "0.3333333333333333 B ['E', 'A', 'B']\n",
      "0.3333333333333333 A ['E', 'B', 'A']\n",
      "1.0 B ['B', 'D', 'E']\n",
      "1.0 B ['B', 'C', 'E']\n",
      "0.0 E ['B', 'A', 'D']\n",
      "0.3333333333333333 C ['A', 'D', 'C']\n",
      "1.0 B ['B', 'A', 'C']\n",
      "1.0 D ['D', 'C', 'B']\n",
      "1.0 A ['A', 'E', 'D']\n",
      "0.3333333333333333 A ['D', 'B', 'A']\n",
      "0.7577777777777777\n"
     ]
    }
   ],
   "source": [
    "if CALC_SCORE:\n",
    "\n",
    "    import numpy as np\n",
    "    def precision_at_k(r, k):\n",
    "        \"\"\"Precision at k\"\"\"\n",
    "        assert k <= len(r)\n",
    "        assert k != 0\n",
    "        return sum(int(x) for x in r[:k]) / k\n",
    "\n",
    "    def MAP_at_3(predictions, true_items):\n",
    "        \"\"\"Score is mean average precision at 3\"\"\"\n",
    "        U = len(predictions)\n",
    "        map_at_3 = 0.0\n",
    "        for u in range(U):\n",
    "            user_preds = predictions[u]\n",
    "            user_true = true_items[u]\n",
    "            user_results = [1 if item == user_true else 0 for item in user_preds]\n",
    "            for k in range(min(len(user_preds), 3)):\n",
    "                map_at_3 += precision_at_k(user_results, k+1) * user_results[k]\n",
    "        return map_at_3 / U\n",
    "\n",
    "    maps = []\n",
    "    for idx, row in test.iterrows():\n",
    "\n",
    "        tl = row[\"answer\"]\n",
    "        pr = row[\"prediction\"].split(\" \")\n",
    "\n",
    "        map = MAP_at_3([pr], [tl])\n",
    "        print(map, tl, pr)\n",
    "        maps.append(map)\n",
    "\n",
    "    print(np.mean(maps))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "0a7578d6",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-10T12:55:10.381627Z",
     "iopub.status.busy": "2023-10-10T12:55:10.380710Z",
     "iopub.status.idle": "2023-10-10T12:55:10.422119Z",
     "shell.execute_reply": "2023-10-10T12:55:10.421133Z"
    },
    "papermill": {
     "duration": 0.174257,
     "end_time": "2023-10-10T12:55:10.423717",
     "exception": false,
     "start_time": "2023-10-10T12:55:10.249460",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>prediction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>A D B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>D C B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>C B E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>C E A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>A D C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>295</td>\n",
       "      <td>A D C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>296</td>\n",
       "      <td>B A C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>297</td>\n",
       "      <td>D C B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>298</td>\n",
       "      <td>A E D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>299</td>\n",
       "      <td>D B A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>300 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      id prediction\n",
       "0      0      A D B\n",
       "1      1      D C B\n",
       "2      2      C B E\n",
       "3      3      C E A\n",
       "4      4      A D C\n",
       "..   ...        ...\n",
       "295  295      A D C\n",
       "296  296      B A C\n",
       "297  297      D C B\n",
       "298  298      A E D\n",
       "299  299      D B A\n",
       "\n",
       "[300 rows x 2 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Once we write our submission file we're good to submit!\n",
    "test[['id', 'prediction']].to_csv('submission.csv', index=False)\n",
    "pd.read_csv(\"submission.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ff4afaa",
   "metadata": {
    "papermill": {
     "duration": 0.125349,
     "end_time": "2023-10-10T12:55:10.675516",
     "exception": false,
     "start_time": "2023-10-10T12:55:10.550167",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 3208.758864,
   "end_time": "2023-10-10T12:55:13.195783",
   "environment_variables": {},
   "exception": null,
   "input_path": "__notebook__.ipynb",
   "output_path": "__notebook__.ipynb",
   "parameters": {},
   "start_time": "2023-10-10T12:01:44.436919",
   "version": "2.4.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
